## Master script for Leo's plots, thesis edition

# load libraries
library(tidyverse)
library(readxl)
library(ComplexHeatmap)
library(circlize)
library(ggrepel)
library(PFun)
library(forcats)
library(FactoMineR) # for PCA
library(factoextra) # for PCA
library(gtools)
library(broom)
library(here)
library(openxlsx)
library(coin)
# library(RVAideMemoire)
library(colorspace)


# options(width = 170)

# set working directory to selected directory
# setwd("/Users/dmarti14/Documents/MRC_Postdoc/Projects/Leo/DATA_TO_PLOT")

# output directory
# dir.create('Summary', showWarnings = TRUE, recursive = FALSE, mode = "0777")





#############################
### 4-way screen (biolog) ###
#############################


### Load data

info = read_csv('Biolog_metabolites.csv',quote = "\"")

info %>% filter(Plate == 'PM5')

# load bacterial data

data.b = read_csv('../Bacteria data/All_files/Data_Leo5/Summary.csv', quote = "\"") %>%
  rename(logAUC_raw = `750nm_f_logAUC`) %>% #`750nm_f_logAUC` data column is what we need for logAUC values
  mutate(Strain = as.character(Strain),
         Strain = recode(Strain, 'BW25113' = 'BW'), #Change strain namings
         ReplicateB = paste0('B', as.character(Replicate)), #Generate replicate ID which indicate that it's data for bacteria
         Type = ifelse(Type == 'Control','C','T'),
         Sample = paste(Strain, Type, sep = '_'),
         Type = factor(Type,
                     levels = c('C', 'T'),
                     labels = c('Control', 'Treatment')),
         Row = str_match_all(Well,'[:digit:]{1,}'), #Get plate row names from well name. 
         Col = str_match_all(Well,'[:alpha:]{1,}'), #Get plate column names from well name
         Row = factor(Row, levels = 1:12), #Make them categorical variable with set order them
         Col = factor(Col, levels = LETTERS[1:8]),
         SampleID = paste(Sample, Replicate, sep = '_'),
         Sample = as.factor(Sample),
         Strain = as.factor(Strain)) %>% #Change Type column coding
  group_by(Strain, Type, Plate, Replicate, Group) %>%
  mutate(logAUC = logAUC_raw - logAUC_raw[Metabolite == 'Negative Control']) %>% # Normalisation agains negative control in each plate
  select(SampleID, Sample, Strain, Type, Replicate, ReplicateB, Index, Plate, 
    Well, Row, Col, Metabolite, MetaboliteU, EcoCycID, KEGG_ID, Group, Class, logAUC_raw, logAUC) %>%
  ungroup 
  

# deleting replicate 3 of pyrE treatment (pyrE_T_3), Biospa failed to deliver data correctly 
data.b = data.b %>%
  filter(SampleID != "pyrE_T_3")


### Summary statistics
# Note: we do not use several of these variables for the plots, but could be useful
# They were generated by Pov 

bac.byrep = data.b %>%
  	unite(STR, Strain, Type, ReplicateB) %>% #Generete unique IDs by concatenating Strain, Type and Replicate values
  	# Select only the columns which are necessary: descriptions, values, exclude Replicate (Now it's ReplicateB), Comment, Reader
  	select(STR, Plate, Well, Row, Col, Index:Class,logAUC_raw, -Replicate) %>% 
  	spread(STR, logAUC_raw) #Spread values into a wide table



# Replicates per plate
data.breps = data.b %>%
  	group_by(Strain, Plate, Type) %>%
  	summarise(Replicates = n()/96)


# Generate summary statistics for each group
bac.sum = data.b %>%
  	group_by(Strain, Plate, Replicate, Group) %>%
  	gather(Measure, Value, logAUC_raw, logAUC) %>%
  	group_by(Measure, Strain, Type, Plate, Well, Index, Metabolite, MetaboliteU) %>% #Work on subsets based on selected grouping variables
  	summarise(Mean = mean(Value, na.rm = TRUE), # Do selected summary over group items - replicates, ignore missing values
            SD = sd(Value, na.rm = TRUE),
            SE = SD/sqrt(n())) %>% # standard error of the mean
  	ungroup %>%
  	mutate(PE = Mean+SE,
         NE = Mean-SE)



# With additional info
bac.sum2 = data.b %>%
  filter(Strain == 'BW' & Plate %in% c('PM1','PM2A', 'PM5')) %>%
  group_by(Strain, Plate, Replicate, Group) %>%
  mutate(logAUC_C = logAUC_raw - logAUC_raw[Metabolite == 'Negative Control' & Type == 'Control']) %>% # Normalisation agains negative control in control
  ungroup %>%
  gather(Measure, Value, logAUC_C, logAUC_raw, logAUC) %>%
  group_by(Measure, Strain, Type, Plate, Well, Index, Metabolite, MetaboliteU) %>% #Work on subsets based on selected grouping variables
  summarise(Mean = mean(Value, na.rm = TRUE), #Do selected summary over group items - replicates, ignore missing values
            SD = sd(Value, na.rm = TRUE),
            SE = SD/sqrt(n())) %>%
  ungroup %>%
  mutate(PE = Mean + SE,
         NE = Mean - SE)




bac.sum.sbs = bac.sum %>% #Remove table grouping. Can be checked via glimpse(bac.sum)
  filter(Measure == 'logAUC_raw') %>%
  select(Strain, Type, Plate, Well, Index, Mean, SD) %>%
  #Transform the table wide to long, Mean and SD become values of a column named Value, while Mean and SD coding becomes Stat column values
  gather(Stat, Value, Mean, SD) %>% 
  unite(STS, Strain, Type, Stat) %>% # Generate unique IDs for summary columns based on Strain, Type and Stat - statistic values
  spread(STS, Value) # Spread the table to wide, showing statistics in each group


bacall = bac.byrep %>% 
  left_join(bac.sum.sbs) %>% 
  select(-c(Index, Metabolite, MetaboliteU)) %>% # Join with raw replicate data
  arrange(Plate, Col, Row) %>%
  select(-c('Row', 'Col'))


# Get timeseries data for the bacterial growth information
# It will take a while
data.b.ts = read_csv('../Bacteria data/All_files/Data_Leo5/Timeseries.csv',quote = "\"") %>%
  filter(Data == '750nm_f') %>% #Select only the relevant data to speed it up
  gather(Time_s, OD, matches('\\d')) %>%
  filter(!is.na(OD)) %>% #Remove empty values if there are missmatches
  mutate(Strain = as.character(Strain),
         Strain = recode(Strain,'BW25113'='BW'), #Change strain namings
         ReplicateB = paste0('B', as.character(Replicate)), #Generate replicate ID which indicate that it's data for bacteria
         Type = ifelse(Type == 'Control','C','T'),
         Type = factor(Type,
                     levels = c('C','T'),
                     labels = c('Control','Treatment')),
         Time_s = as.numeric(Time_s),
         Time_h = Time_s/3600,
         Row = str_match_all(Well,'[:digit:]{1,}'), #Get plate row names from well name. 
         Col = str_match_all(Well,'[:alpha:]{1,}'), #Get plate column names from well name
         Row = factor(Row, levels = 1:12), #Make them categorical variable with set order them
         Col = factor(Col, levels = LETTERS[1:8])) #Make them categorical variable with set order them


data.b.ts = data.b.ts %>% mutate(OD = as.numeric(OD))


# Growth curves for summary

tsum = data.b.ts %>%
  group_by(Strain, Type, Index, Plate, Well, Metabolite, MetaboliteU, Time_h) %>%
  summarise(Mean = mean(OD), SD = sd(OD),SE = SD/sqrt(length(OD))) %>%
  ungroup 


### load worm data


# Let's organise worm data
# Worm data
data.bw = read_xlsx('Biolog_worm/BW/BW_summary.xlsx', sheet = 'Summary')
data.crp = read_xlsx('../Worm_data/Biolog Worm nutrient screen - CRP.xlsx',sheet= 'All')
data.pyrE = read_xlsx('Biolog_worm/pyrE/pyrE_summary.xlsx',sheet = 'Summary')
data.TM = read_xlsx('Biolog_worm/udp udk upp/TM_summary.xlsx',sheet = 'Summary')

# Join all data, create scores, plates, experiments and worm replicate names

data.wa = rbind(data.bw, data.crp, data.pyrE, data.TM) %>%
  gather(Column, Worm_phenotype,`1`:`12`) %>%
  mutate(Score = as.numeric(ifelse(grepl('[0-9]', Worm_phenotype), as.numeric(Worm_phenotype), NA)),
         Plate = recode(Plate, 'PM2'='PM2A', 'PM3'='PM3B', 'PM4'='PM4A'),
         Replicate = as.character(Replicate),
         ReplicateW = paste0('W',Replicate),
         Experiment = paste0(ifelse(Type == 'Control', 'C','T'),`5FU_uM`)) %>%
  unite(Well, Row, Column, sep = '') %>%
  unite(SER, Strain, Experiment, ReplicateW, remove = FALSE) %>%
  left_join(info) %>%
  drop_na(Strain) %>%
  select(Strain, Type,`5FU_uM`, Experiment, Replicate, ReplicateW, SER, Plate, Well, Index:Class, Worm_phenotype, Score) 


### IMPORTANT DATA MODIFICATION:
  # everything lower or equal than 1 is going to be equal to 1

# data.wa[data.wa$Score <= 1 & !is.na(data.wa$Score),]$Score = 1
# data.wa[data.wa$Worm_phenotype <= 1 & !is.na(data.wa$Worm_phenotype),]$Worm_phenotype = 1


# Worm raw data side by side comparison
worm.byrep = data.wa %>%
  #left_join(info[,c('Plate','Well','MetaboliteU')]) %>%
  select(Plate, Well, Index, Metabolite, MetaboliteU, EcoCycID, KEGG_ID, SER, Score) %>%
  spread(SER, Score)
 

worm.sum = data.wa %>%
  group_by(Strain, Plate, Well, Experiment) %>%
  summarise(Median = median(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            N = sum(!is.na(Score)),
            SD = sd(Score, na.rm = TRUE))
  

worm.sum.clean = worm.sum %>%
  gather(Stat, Value, Median:SD) %>%
  unite(SES, Strain, Experiment, Stat) %>%
  filter(!( SES == 'CRP_C0_SD' |
            SES == 'pyrE_C0_SD' 
            ) ) %>%
  select(Plate, Well, SES, Value) %>%
  spread(SES, Value)

## this creates a HUGE table with 64 variables 
wormall = worm.byrep %>%
  left_join(worm.sum.clean)


###
### PCA plots
###

# this big and dummy function will plot the PCA of the selected strains and save them in a PDF file
pca_plot = function(strain = "BW"){
  # filter by strain and Plate if you want
  pca_b_data = data.b %>%
    filter(Strain %in% c(strain) & Metabolite != 'Negative Control')

  # some data frame transformations
  pca_b_data = pca_b_data %>%
    select(SampleID, MetaboliteU, logAUC) %>%
    spread(MetaboliteU, logAUC) %>%
    data.frame(check.names = F)

  rownames(pca_b_data) = pca_b_data[,1]; pca_b_data = pca_b_data[,-1]

  # lets compute the PCA
  res.pca = PCA(pca_b_data, scale.unit = TRUE, ncp = 5, graph = F)

  # metadata 
  meta_var = bioinfo %>% filter(Strain %in% c(strain))

  # extract info about the individuals
  ind = get_pca_ind(res.pca)
  ind_df = data.frame(ind$coord[,1], ind$coord[,2], ind$coord[,3], meta_var$Type, 
    meta_var$Strain)

  colnames(ind_df) = c('Dim1', 'Dim2', 'Dim3', 'Type', 'Strain')

  # get ellipses based on the correlation
  getellipse = function(x, y, sc = 1) {
    as.data.frame(ellipse::ellipse(cor(x, y),
                                    scale = c(sd(x) * sc, sd(y) * sc),
                                    centre = c(mean(x), mean(y))))
  }

  # make a data frame from ellipses
  ell = ind_df %>% group_by(Type, Strain) %>% do(getellipse(.$Dim1, .$Dim2, 1)) %>% data.frame

  # plot!
  ggplot(ind_df, aes(x = Dim1, y = Dim2, color = Strain, group = interaction(Type, Strain))) + 
    geom_point(size = 3, show.legend = NA, alpha = 0.5) + 
    geom_path(data = ell, aes(x = x, y = y, group = interaction(Type, Strain), linetype = Type), size = 1) +
    geom_polygon(data = ell, aes(x = x, y = y, group = interaction(Type, Strain), 
      linetype = Type, fill = Strain), size = 1, alpha = 0.3) +
    xlab(paste("PC1 - ", round(res.pca$eig[1,2], 1), " % of variance", sep = "")) + 
    ylab(paste("PC2 - ", round(res.pca$eig[2,2], 1), " % of variance", sep = "")) +
    ggtitle(paste("PCA of strain:", unique(ell$Strain))) +
    theme(plot.title = element_text(hjust = 0.5),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank())

  # save the file
  ggsave(file = here('Summary',paste('PCA_', strain , '.pdf',sep = '')),
  width = 100, height = 80, units = 'mm', scale = 2, device = cairo_pdf, family = "Arial")
}

pca_plot("BW")
pca_plot("TM")
pca_plot("pyrE")
pca_plot(c("crp", "BW", "TM", "pyrE"))
# change name of generated file before next line
pca_plot("crp")





###########################
#### Time series 


# Get timeseries data
# It will take a while
data.b.ts = read_csv('../Bacteria data/All_files/Data_Leo5/Timeseries.csv',quote = "\"") %>%
  filter(Data == '750nm_f') %>% #Select only the relevant data to speed it up
  gather(Time_s, OD, matches('\\d')) %>%
  filter(!is.na(OD)) %>% #Remove empty values if there are missmatches
  mutate(Strain = as.character(Strain),
         Strain = recode(Strain,'BW25113'='BW'), #Change strain namings
         ReplicateB = paste0('B', as.character(Replicate)), #Generate replicate ID which indicate that it's data for bacteria
         Type = ifelse(Type == 'Control','C','T'),
         Type = factor(Type,
                     levels = c('C','T'),
                     labels = c('Control','Treatment')),
         Time_s = as.numeric(Time_s),
         Time_h = Time_s/3600,
         Row = str_match_all(Well,'[:digit:]{1,}'), #Get plate row names from well name. 
         Col = str_match_all(Well,'[:alpha:]{1,}'), #Get plate column names from well name
         Row = factor(Row, levels = 1:12), #Make them categorical variable with set order them
         Col = factor(Col, levels = LETTERS[1:8])) #Make them categorical variable with set order them

# [:digit:]{1,} - one or more digits
# This is a 'regular expression'
# More info here https://stringr.tidyverse.org/articles/regular-expressions.html

data.b.ts = data.b.ts %>% mutate(OD = as.numeric(OD))



# Growth curves for summary

tsum = data.b.ts %>%
  group_by(Strain, Type, Index, Plate, Well, Metabolite, MetaboliteU, Time_h) %>%
  summarise(Mean = mean(OD), SD = sd(OD),SE = SD/sqrt(length(OD))) %>%
  ungroup 


Metcols = c("black", "red")
names(Metcols) = c("Control","Treatment")
Metlab = 'Type'



mlevels = c('Negative Control_C','Uridine_N','L-Serine','Adenosine')

mlabels = c('Control','Uridine','L-Serine','Adenosine')

# This may take a while
# plot growth curves by different conditions


tsum %>%
  filter(Strain == 'BW') %>%
  filter(MetaboliteU %in% mlevels) %>%
  mutate(MetaboliteU = factor(MetaboliteU, levels = mlevels, labels = mlabels)) %>%
  ggplot(aes(x = Time_h, y = Mean, fill = Type, color = Type)) +
  geom_ribbon(aes(ymin = Mean - SD, ymax = Mean + SD), color = NA, alpha = 0.2) +
  geom_line() +
  scale_x_continuous(breaks = seq(0, 24, by = 12)) +
  ylab("OD") +
  xlab("Time, h") +
  labs(fill="Type") +
  #facet_wrap(~MetaboliteU,ncol = 2)+
  facet_wrap(~MetaboliteU,ncol = 2) +
  scale_colour_manual(name = Metlab, values = Metcols) +
  scale_fill_manual(name = Metlab, values = Metcols) +
  theme(legend.position = "top") +
  theme_light()

## 

# tsum %>%
#   filter(MetaboliteU %in% mlevels) %>%
#   mutate(MetaboliteU = factor(MetaboliteU, levels = mlevels, labels = mlabels)) %>%
#   mutate(Strain = factor(Strain)) %>%
#   ggplot( aes(x = Time_h, y = Mean, fill = Type, color = Type)) +
#   geom_ribbon(aes(ymin = Mean - SD, ymax = Mean + SD), color = NA, alpha = 0.2) +
#   geom_line() +
#   scale_x_continuous(breaks = seq(0, 24, by = 12)) +
#   ylab("OD") +
#   xlab("Time, h") +
#   labs(fill = "Type") +
#   #facet_wrap(~MetaboliteU,ncol = 2)+
#   facet_grid((MetaboliteU ~ Strain)) +
#   scale_colour_manual(name = Metlab, values = Metcols) +
#   scale_fill_manual(name = Metlab, values = Metcols) +
#   theme(legend.position = "top", strip.text = element_text(size = 13))



quartz.save(file = here('Summary', '4_way_growth_curves.pdf'),
    type = 'pdf', dpi = 300, height = 8, width = 10)


############################


# 4-way plot --------------------------------------------------------------



###
### 4-way plot
###

# is a hell of a code, it's super messy, but it works...

worm.sum[worm.sum$Strain == 'CRP', ]$Strain = 'crp'
worm.sum$Strain = as.factor(worm.sum$Strain)



# print the list of variables 
for (name in unique(worm.sum$Strain)){
  print(name)
  print(unique((worm.sum %>% filter(Strain == name))$Experiment))
}

# select only one experiment and strain
# contrasts and other main variables of this part

strain = 'BW'
experiment = "T5"


str_C = paste(strain, '_C', sep = '')
str_T = paste(strain, '_T', sep = '')
str_CT = paste(strain, '_5FU', sep = '')
alln = paste(strain, '_5FU_Alln', sep = '')
adjs = paste(strain, '_5FU_adj', sep = '')


# for example, to mimic Pov's code
worm.sum %>% filter(Experiment == experiment, Strain == strain)

worm.old = worm.sum %>%
  filter(Experiment == experiment & Strain == strain) %>% 
  ungroup %>%
  mutate(Description = 'C. elegans development at 5uM 5-FU',
         Contrast = 'Ce_Dev5',   # C. elegans developement with 5FU
         Contrast_type = 'Treatment',
         FDR = NA) %>%
  select(Description, Contrast, Contrast_type, Plate, Well, logFC = Median, SE = SD, FDR)

# check that we filtered stuff properly
worm.old %>%
  group_by(Plate) %>%
  summarise(N = n())


# Join bacterial and worm results (change contrasts)
jointresults = allresults$results %>%
  filter((Contrast %in% c(str_C, str_T, str_CT, alln))) %>%
  mutate(Contrast = fct_recode(Contrast, adjs = alln)) %>% 
  select(Description, Contrast, Contrast_type, Plate, Well, logFC, SE, FDR) %>%
  bind_rows(worm.old) %>%
  left_join(info) %>%
  select(Description:Well,Index:KEGG_ID,logFC:FDR)


jointcast = jointresults %>%
  select(Contrast, Plate, Well, Index, Metabolite, MetaboliteU, EcoCycID, KEGG_ID, logFC, SE, FDR) %>%
  gather(Stat, Value, logFC, SE, FDR) %>%
  unite(CS, Contrast, Stat) %>%
  spread(CS, Value) %>%
  rename(Ce_Dev5_Median = Ce_Dev5_logFC, Ce_Dev5_SD = Ce_Dev5_SE) %>%
  select(-Ce_Dev5_FDR)


# write.csv(jointresults,paste(odir,'/Ecoli_results_All_As_old_screen.csv', sep = ''),row.names = FALSE)
# write.csv(jointcast,paste(odir,'/Ecoli_results_sidebyside_As_old_screen.csv', sep = ''),row.names = FALSE)

 
# Multiplex 
jointresults.multi = multiplex(jointresults,c("Plate","Well","Index","Metabolite","MetaboliteU","EcoCycID","KEGG_ID"))
jointresults.multi2 = multiplex(jointresults,c("Plate","Well","Index","Metabolite","MetaboliteU","EcoCycID","KEGG_ID"),2)




########################
NGMa = adjustments[adjustments$Contrast == alln, ]$a
NGMb = adjustments[adjustments$Contrast == alln, ]$b



gradcolours = c('#71B83B','yellow','orange','red')




total = jointresults.multi %>%
  filter(x_Contrast == str_C & y_Contrast == str_T & z_Contrast == 'Ce_Dev5')

first = jointresults.multi %>%
  filter(x_Contrast == str_C & y_Contrast == str_T & z_Contrast == 'Ce_Dev5') %>%
  filter(z_logFC <=1)

second = jointresults.multi %>%
  filter(x_Contrast == str_C & y_Contrast == str_T & z_Contrast == 'Ce_Dev5') %>%
  filter(z_logFC >1 & z_logFC <= 2)

third = jointresults.multi %>%
  filter(x_Contrast == str_C & y_Contrast == str_T & z_Contrast == 'Ce_Dev5') %>%
  filter(z_logFC >2 & z_logFC <= 3)

fourth = jointresults.multi %>%
  filter(x_Contrast == str_C & y_Contrast == str_T & z_Contrast == 'Ce_Dev5') %>%
  filter(z_logFC >3 & z_logFC <= 4)


ggplot(first, aes()) +
  geom_abline(intercept = 0, slope = 1, alpha = 0.3, color = 'grey', linetype = 'longdash') +
  geom_abline(aes(intercept = NGMb, slope = NGMa), alpha = 0.6, color = 'red') +
  geom_vline(aes(xintercept = 0), alpha = 0.9, color = 'grey') +
  geom_hline(aes(yintercept = 0), alpha = 0.9, color = 'grey') +
  geom_errorbarh(data = total, aes(y = y_logFC, xmax = x_logFC + x_SE, xmin = x_logFC - x_SE), height = 0, alpha = 0.3, color = 'grey50') +
  geom_errorbar(data = total, aes(x = x_logFC, ymax = y_logFC + y_SE, ymin = y_logFC - y_SE), width = 0, alpha = 0.3, color = 'grey50') +
  geom_point(aes(x = x_logFC, y = y_logFC, colour = z_logFC), alpha = 0.85, size = 2) +
  geom_point(data = second, aes(x = x_logFC, y = y_logFC, colour = z_logFC), alpha = 0.85, size = 2) +
  geom_point(data = third, aes(x = x_logFC, y = y_logFC, colour = z_logFC), alpha = 0.85, size = 2) +
  geom_point(data = fourth, aes(x = x_logFC, y = y_logFC, colour = z_logFC), alpha = 0.85, size = 2) +
  scale_size(range = c(1,5), name = 'C. elegans\nphenotype') +
  scale_colour_gradientn(colours = gradcolours,
                         breaks = c(1,2,3,4), limits = c(1,4), guide = "legend", name = 'C. elegans\nphenotype') +
  scale_y_continuous(breaks = -5:5)+
  coord_cartesian(xlim = c(-3, 3), ylim = c(-3, 4)) +
  labs(title = paste(strain, " growth with 5FU treatment at 0 uM", sep = ''),
  x = expression(paste(italic("E. coli"), ' growth vs NGM - Control, logFC', sep = '')), 
  y = expression(paste(italic('E. coli'), ' growth vs NGM - 5-FU Treatment, logFC', sep = ''))) +
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
  # geom_text_repel(data = fourth, aes(x = x_logFC, y = y_logFC, label = ifelse(MetaboliteU %in% mlevels, MetaboliteU, '')), size = 4, box.padding = unit(1.6, "lines")) + # USE ONLY FOR DEMONSTRATIVE PURPOSES FOR LEO
  geom_text_repel(data = fourth, aes(x = x_logFC, y = y_logFC, label = ifelse(z_logFC >= 4, MetaboliteU, '')), box.padding = unit(0.6, "lines"), segment.alpha = 0.4) + # only name those nutr with C. elegans phenotype 3 or more
  geom_text_repel(data = third, aes(x = x_logFC, y = y_logFC, label = ifelse(z_logFC >= 3, MetaboliteU, '')), box.padding = unit(0.6, "lines"), segment.alpha = 0.4) + # only name those nutr with C. elegans phenotype 3 or more
  # geom_text_repel(aes(label = ifelse(z_logFC >= 3 | z_logFC == 0, MetaboliteU, '')), box.padding = unit(0.6, "lines"), segment.alpha = 0.4) + # only name those nutr with C. elegans phenotype 3 or more
  guides(color = guide_legend()) +
  theme(panel.grid.major = element_line(size = 0.06, colour = "grey50"),
      # panel.grid.major.y = element_blank(),
      panel.grid.minor.y = element_blank(),
      panel.background = element_rect(fill = "white", colour = "grey50"),
      legend.text = element_text(size = 15),
      legend.title = element_text(size = 15),
      axis.text.x = element_text(size = 10),
      axis.text.y = element_text(size = 10)) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  annotate(geom = 'text', x = -2, y = 1.5, label = paste('y = ', round(NGMa,3), 'x +', round(NGMb,3), sep = ''), size = 5)


quartz.save(file = here('Summary', paste0("Scatter_", strain, "_5uM.pdf")),
	type = 'pdf', dpi = 300, height = 8, width = 10)


# quartz.save(file = here('Summary', paste0("Scatter_EXAMPLE_", strain, "_5uM.pdf")),
#     type = 'pdf', dpi = 300, height = 8, width = 10)


geom_text_repel(aes(label = ifelse(Genes %in% c('Δglta ΔprpB::K','ΔgltA::K'), as.character(Genes), '')), position = pos, colour = 'red', size = 5, box.padding = 3.5) +

# scales for different mutants:
# BW: coord_cartesian(xlim = c(-5, 2), ylim = c(-4, 4))
# pyrE: coord_cartesian(xlim = c(-5, 2), ylim = c(-4, 4))
# TM: coord_cartesian(xlim = c(-4, 2), ylim = c(-4, 2.5))








####
# linear modeling, statistical analyses



contrasts = read.contrasts('!Contrasts.xlsx') # This table describes comparisons made in data

# loads the metadata table
contrasts.table = contrasts$Contrasts.table
# loads the contrasts matrix (1s and 0s)
contr.matrix = contrasts$Contrasts.matrix

# shortens the metadata data frame
contr.desc = contrasts.table %>%
  select(Description:Strain)



# Initialise adjustments
group = c('All','PM4A','PM3B','PM2A','PM1')
strain = c('BW','TM','crp','pyrE')


adjustments = expand.grid('Group' = group, 'Strain' = strain) %>%
  mutate(a = -1, b = 0,
         Contrast = paste0(Strain, '_5FU_', Group, 'n')) %>%
  select(Contrast, Group:b) 



# transform the contr.matrix into a real matrix (not data frame)
contr.matrix = contr.matrix %>%
  as.matrix()


# this piece of code performs a lot of different things, be careful

allresults = data.b %>%
  #filter(Strain == 'BW') %>%
  filter(!Metabolite %in% c('Negative Control','Positive Control') ) %>%
  # filter(! (Plate == 'PM3B' & SampleID == 'BW_C_1') ) %>%
  # filter(! (Plate == 'PM4A' & SampleID == 'TM_C_3') ) %>%
  group_by(Plate, Well, Index, Metabolite, MetaboliteU, EcoCycID, KEGG_ID) %>%
  # Another creation of mine. Nothing fancy, just a wrapper for the use of contrasts in LM. 
  # But integrates nicely with tidyverse workflow. Code in PFun
  do(hypothesise(.,"logAUC~0+Sample", contr.matrix)) %>% 
  group_by(Plate, Well, Metabolite, MetaboliteU, EcoCycID, KEGG_ID) %>%
  getresults(contr.desc) # Calculate some additional parameters. Code in PFun

#get results in different shapes
results = allresults$results
results.cast = allresults$cast
results.castfull = allresults$castfull


results.cor = multiplex(results,c("Strain", "Plate", "Well", "Index", "Metabolite", "MetaboliteU", "EcoCycID", "KEGG_ID"), 2) %>% 
  filter(str_detect(x_Contrast,'_C') &
           str_detect(y_Contrast,'_T'))

# Step No 2
# Find contrast adjustments
confounding = function(adjustments, data) {
  
  grp = as.character(adjustments$Group)
  strn <= as.character(adjustments$Strain)


  print(paste0("Strain: ",strn) )
  print(paste0("Group: ",grp) )
  
  seldata=data %>%
    filter(Strain==strn)
  
  if (grp=='All'){
    seldata=seldata
  } else if (grp=='PM1'){
    seldata=seldata %>%
      filter(Plate %in% c('PM1'))
  } else if (grp=='PM3B') {
    seldata=seldata %>%
      filter(Plate %in% c('PM3B'))
  } else if (grp=='PM4A') {
    seldata=seldata %>%
      filter(Plate %in% c('PM4A'))
  } else if (grp=='PM2A') {
    seldata=seldata %>%
      filter(Plate %in% c('PM2A'))
  } 

  print(paste0('Plates: ', paste0(seldata$Plate %>% unique %>% print,collapse=';') ) )
  print(paste0('Rows: ',dim(seldata)[1]) )
  
  
  if (dim(seldata)[1] == 0){
    result=data.frame('a' = NA,
                       'neg_a' = NA,
                       'b' = NA,
                       'a_p' = NA,
                       'b_p' = NA,
                       'r2' = NA)
    return(result)
  }

  print('First round')
  
  # print(unique(seldata$x_Contrast))
  # print(unique(seldata$y_Contrast))
  
  # frml=paste0(strn,"_T_logFC~",strn,"_C_logFC")
  # print(paste0("Formula: ",frml) )
  
  #genfit=lm(seldata[,paste0(strn,"_T_logFC")]~seldata[,paste0(strn,"_C_logFC")])
  
  #attach(seldata)
  attach(seldata)
  genfit = lm("y_logFC~x_logFC")
  
  genres = summary(genfit)
  
  car::outlierTest(genfit) # Automatically remove outliers, which could skew the correlation. Especially valuable in the case of 5-FU
  ot = car::outlierTest(genfit)
  car::qqPlot(genfit, main = "QQ Plot")
  outlist=c(names(ot$rstudent))
  #Outliers:
  print(paste('Outliers in ', grp, sep = ''))
  print(seldata[outlist,])
  #
  #
  filtseldata = seldata[setdiff(rownames(seldata),outlist),]
  
  detach(seldata)

  print('Second round')
  # 
  #Get final fit
  attach(filtseldata)
  genfit2 = lm("y_logFC~x_logFC")
  genres = summary(genfit2)
  
  result = data.frame('a' = genres$coefficients[[2]],
                     'neg_a' = -genres$coefficients[[2]],
                     'b' = genres$coefficients[[1]],
                     'a_p' = genres$coefficients[[8]],
                     'b_p' = genres$coefficients[[7]],
                     'r2' = genres$r.squared)
  

  return(result)

}



# Update adjustments based on identified correlations. Go back to line 806 and rerun linear modeling with adjustments
adjustments = adjustments %>%
  group_by(Contrast,Group,Strain) %>%
  do(confounding(adjustments = ., data = results.cor)) #



# Update this based on identified adjustment values
# Only run this after the first round of analysis
#######################################


contr.matrix = contr.matrix %>%
  data.frame()


contr.matrix['BW_5FU_Alln',c('BW_C','m')] = adjustments[adjustments$Contrast ==  'BW_5FU_Alln', c('neg_a','b')]
contr.matrix['TM_5FU_Alln',c('TM_C','m')] = adjustments[adjustments$Contrast ==  'TM_5FU_Alln', c('neg_a','b')]
contr.matrix['crp_5FU_Alln',c('crp_C','m')] = adjustments[adjustments$Contrast == 'crp_5FU_Alln', c('neg_a','b')]
contr.matrix['pyrE_5FU_Alln',c('pyrE_C','m')]  = adjustments[adjustments$Contrast == 'pyrE_5FU_Alln', c('neg_a','b')]


#######################################


#### new round of statistical analyses
contr.matrix = contr.matrix %>%
  as.matrix()
# this piece of code performs a lot of different things, be careful

allresults = data.b %>%
  #filter(Strain == 'BW') %>%
  filter(!Metabolite %in% c('Negative Control','Positive Control') ) %>%
  # filter(! (Plate == 'PM3B' & SampleID == 'BW_C_1') ) %>%
  # filter(! (Plate == 'PM4A' & SampleID == 'TM_C_3') ) %>%
  group_by(Plate, Well, Index, Metabolite, MetaboliteU, EcoCycID, KEGG_ID) %>%
  # Another creation of mine. Nothing fancy, just a wrapper for the use of contrasts in LM. 
  # But integrates nicely with tidyverse workflow. Code in PFun
  do(hypothesise(.,"logAUC~0+Sample", contr.matrix)) %>% 
  group_by(Plate, Well, Metabolite, MetaboliteU, EcoCycID, KEGG_ID) %>%
  getresults(contr.desc) # Calculate some additional parameters. Code in PFun



#get results in different shapes
results = allresults$results
results.cast = allresults$cast
results.castfull = allresults$castfull


# Save statistical analysis results

list_of_datasets = list('results' = results, 'results.cast' = results.cast, "results.castfull" = results.castfull)

write.xlsx(list_of_datasets, here('Summary', '4way_stats.xlsx'), colNames = T, rowNames = F) 




#######################################################
### Enrichment analysis of metabolites and pathways ###
#######################################################

# variables to use for this analysis
results
data.wa
info

# load other helper files for KEGG and EcoCyc enrichment
# KEGG mappings
kegg.mappings = read_csv(here('Bacteria data',"All_KEGG_mappings_Complete.csv")) %>%
  separate_rows(all.mapped) %>%
  filter(!is.na(all.mapped)) %>%
  group_by(all.mapped) %>%
  summarise(Pathway_IDs = paste(unique(PathwayID), collapse = ';'),
            Pathways = paste(unique(Description), collapse = ';'))

#EcoCyc mappings
EC_mappings = read_csv(here('Bacteria data','EcoCyc_Class_mappings.csv'))

# filter some results that are not useful now
res = results %>% filter(Contrast_type == 'Treatment')

# some strange thing
ecomulti = multiplex(res, c("Plate","Well","Index","MetaboliteU","EcoCycID","KEGG_ID"),3)

# 
all.results.KEGG = res %>%
  left_join(kegg.mappings,by = c('KEGG_ID'='all.mapped')) %>%
  left_join(EC_mappings) 
  # %>%
  # select(Organism,Organism_short,OMC:KEGG_ID,Pathway_IDs,Pathways,EcoCyc_Instances,EcoCyc_Classes,everything())




EcoCyc.enrichment = all.results.KEGG %>%
  left_join(EC_mappings) %>%
  left_join(info) %>%
  filter(!is.na(FDR) & !is.na(EcoCyc_Classes)) %>%
  enrichment(
             featureid = 'EcoCyc_Instances',
             groups = c('Contrast','Strain', 'Class')) 




EcoCyc.enrichment %>%
  filter(Type!="All" & OMC %in% c("Ec_G_T-C","Ce_F_T-C") & !Class %in% EChm$Class)%>%
  select(-c(NoInstances:Comparison),-c(Organism_short,Measure,Contrast,OMC) ) %>%
  write_csv(paste0(odir,'/EcoCyc_enrichment.csv'))



#---------------------------------------------
#---------------------------------------------
#---------------------------------------------


### using Goseq

# load EC classes
EC_classes = read_csv(here('Bacteria data',"Ecoli_All_class_instances_non_redundant.csv")) 
EC_classes = EC_classes[,-1]


res = results %>% filter(Contrast_type == 'Treatment')

res.BW = res %>% filter(Strain == 'BW', Contrast == 'BW_5FU')
# remove the NAs
resdat = res.BW[complete.cases(res.BW$FDR),]


degenes = as.integer(resdat$FDR<0.05)
# degenes = as.integer(resdat$FDR<0.05 & abs(resdat$log2FoldChange) > 1)
names(degenes) = resdat$EcoCycID

# remove duplicate gene names
degenes = degenes[match(unique(names(degenes)),names(degenes))]

go.class = EC_classes %>% select(Class, Instance)
# go.class = info %>% filter(Plate %in% c('PM1', 'PM2A', 'PM3B', 'PM4A')) %>% select(EcoCycID, Class)


# goseq::goseq(degenes, gene2cat = go.class, method = 'Hypergeometric')


### manual enrichment analysis

# total metab
met = res.BW %>% filter(Metabolite != 'Negative Control') %>% select(EcoCycID) %>% t %>% as.vector
sig.met = res.BW %>% filter(Metabolite != 'Negative Control', FDR < 0.05) %>% select(EcoCycID) %>% t %>% as.vector


go.class = EC_classes %>% 
    filter(NoInstances > 2) 


# go.class = go.class %>%
#     filter(Instance %in% met)


# lets try this class = |Sugar-alcohols|


class.met = go.class %>% filter(Class == '|Sugar-alcohols|') %>% select(Instance) %>% t %>% as.vector

N = length(met)
m = length(class.met)
n = N - m
k = length(sig.met)
x = length(class.met[class.met %in% sig.met])

phyper(q = x - 1, m = m, n = n, k = k, lower.tail = FALSE)



classes = unique(go.class$Class)
dummy = c()
for (class in classes){
  class.met = go.class %>% filter(Class == class) %>% select(Instance) %>% t %>% as.vector
  n = N - m
  k = length(sig.met)
  x = length(class.met[class.met %in% sig.met])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  dummy = c(dummy, fit)
}


names(dummy) = classes

dummy[dummy < 0.05]


met %in% go.class$Instance











#--------------------------------------------------------#







### ok, summary until this point of the code. I got the hypergeometric test to work with our 
### samples, which is good. I would need to make it a function. But, it seems that the database
### I was using is not good, it does not contain all the names Im interested, so I need to make
### a good one from another file from Pov's folders

# read file
EC_classes = read_csv(here('Bacteria data',"All_results_side_by_side_withKEGGEcoCycClass.csv")) %>%
    select(Plate:EcoCyc_Classes) 


# separate rows with strsplit 
# EC_classes %>% 
#     mutate(EcoCyc_Classes = strsplit(as.character(EcoCyc_Classes), ';')) %>% 
#     unnest(EcoCyc_Classes)

# v2, with separate_rows
EC_classes = EC_classes %>% 
    separate_rows(EcoCyc_Classes, sep = ';')

# test if all metabolites are included in the list
met %in% EC_classes$EcoCycID



EC_path = read_csv(here('Bacteria data',"All_results_side_by_side_withKEGGEcoCycClass.csv")) %>%
    select(Plate:EcoCyc_Classes) %>%
    separate_rows(Pathways, sep = ';')



met = res.BW %>% filter(Metabolite != 'Negative Control') %>% select(EcoCycID) %>% t %>% as.vector
class.met = EC_classes %>% filter(EcoCyc_Classes == '|Sugar-alcohols|') %>% select(EcoCycID) %>% t %>% as.vector



N = length(met %in% EC_classes$EcoCycID) # total marked elements
m = length(class.met) # marked elements from a specific category in the total population
n = N - m # non-marked elements
k = length(sig.met) # size of the selection
x = length(class.met[class.met %in% sig.met]) # marked elements from the selection

phyper(q = x - 1, m = m, n = n, k = k, lower.tail = FALSE)




classes = unique(EC_classes$EcoCyc_Classes)
dummy = c()
for (class in classes){
  class.met = EC_classes %>% filter(EcoCyc_Classes == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met)
  x = length(class.met[class.met %in% sig.met])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  dummy = c(dummy, fit)
}

dummy = p.adjust(dummy, method = 'fdr')

names(dummy) = classes

dummy[dummy < 0.05]

###
# let's do it with every set

res.BW = res %>% filter(Strain == 'BW', Contrast == 'BW_5FU')
res.pyrE = res %>% filter(Strain == 'pyrE', Contrast == 'pyrE_5FU')
res.TM = res %>% filter(Strain == 'TM', Contrast == 'TM_5FU')


# total metab
met = res.BW %>% filter(Metabolite != 'Negative Control') %>% select(EcoCycID) %>% t %>% as.vector

# significative metabolites 
sig.met.BW.up =   res.BW %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC > 0) %>% select(EcoCycID) %>% t %>% as.vector
sig.met.BW.down = res.BW %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC < 0) %>% select(EcoCycID) %>% t %>% as.vector

sig.met.pyrE.up =   res.pyrE %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC > 0) %>% select(EcoCycID) %>% t %>% as.vector
sig.met.pyrE.down = res.pyrE %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC < 0) %>% select(EcoCycID) %>% t %>% as.vector

sig.met.TM.up =   res.TM %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC > 0) %>% select(EcoCycID) %>% t %>% as.vector
sig.met.TM.down = res.TM %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC < 0) %>% select(EcoCycID) %>% t %>% as.vector

















### initialize variables

N = length(met %in% EC_classes$EcoCycID) # total marked elements
classes = unique(EC_classes %>% filter(Plate != 'extra') %>% select(EcoCyc_Classes) %>% t %>% as.character())
### calculate enrichment values
### BW
# up 
enrich.BW.up = c()
for (class in classes){
  class.met = EC_classes %>% filter(EcoCyc_Classes == class) %>% select(EcoCycID) %>% t %>% as.vector
  m = length(class.met)
  n = N - m
  k = length(sig.met.BW.up)
  x = length(class.met[class.met %in% sig.met.BW.up])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.BW.up = c(enrich.BW.up, fit)
}

enrich.BW.up = p.adjust(enrich.BW.up, method = 'fdr')
names(enrich.BW.up) = classes
enrich.BW.up = enrich.BW.up[enrich.BW.up < 0.05]

# down 
enrich.BW.down = c()
for (class in classes){
  class.met = EC_classes %>% filter(EcoCyc_Classes == class) %>% select(EcoCycID) %>% t %>% as.vector
  m = length(class.met)
  n = N - m
  k = length(sig.met.BW.down)
  x = length(class.met[class.met %in% sig.met.BW.down])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.BW.down = c(enrich.BW.down, fit)
}

enrich.BW.down = p.adjust(enrich.BW.down, method = 'fdr')
names(enrich.BW.down) = classes
enrich.BW.down = enrich.BW.down[enrich.BW.down < 0.05]





### pyrE
# up
enrich.pyrE.up = c()
for (class in classes){
  class.met = EC_classes %>% filter(EcoCyc_Classes == class) %>% select(EcoCycID) %>% t %>% as.vector
  m = length(class.met)
  n = N - m
  k = length(sig.met.pyrE.up)
  x = length(class.met[class.met %in% sig.met.pyrE.up])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.pyrE.up = c(enrich.pyrE.up, fit)
}

enrich.pyrE.up = p.adjust(enrich.pyrE.up, method = 'fdr')
names(enrich.pyrE.up) = classes
enrich.pyrE.up = enrich.pyrE.up[enrich.pyrE.up < 0.05]

# down
enrich.pyrE.down = c()
for (class in classes){
  class.met = EC_classes %>% filter(EcoCyc_Classes == class) %>% select(EcoCycID) %>% t %>% as.vector
  m = length(class.met)
  n = N - m
  k = length(sig.met.pyrE.down)
  x = length(class.met[class.met %in% sig.met.pyrE.down])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.pyrE.down = c(enrich.pyrE.down, fit)
}

enrich.pyrE.down = p.adjust(enrich.pyrE.down, method = 'fdr')
names(enrich.pyrE.down) = classes
enrich.pyrE.down = enrich.pyrE.down[enrich.pyrE.down < 0.05]







### TM
# up
enrich.TM.up = c()
for (class in classes){
  class.met = EC_classes %>% filter(EcoCyc_Classes == class) %>% select(EcoCycID) %>% t %>% as.vector
  m = length(class.met)
  n = N - m
  k = length(sig.met.TM.up)
  x = length(class.met[class.met %in% sig.met.TM.up])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.TM.up = c(enrich.TM.up, fit)
}

enrich.TM.up = p.adjust(enrich.TM.up, method = 'fdr')
names(enrich.TM.up) = classes
enrich.TM.up = enrich.TM.up[enrich.TM.up < 0.05]


# down
enrich.TM.down = c()
for (class in classes){
  class.met = EC_classes %>% filter(EcoCyc_Classes == class) %>% select(EcoCycID) %>% t %>% as.vector
  m = length(class.met)
  n = N - m
  k = length(sig.met.TM.down)
  x = length(class.met[class.met %in% sig.met.TM.down])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.TM.down = c(enrich.TM.down, fit)
}

enrich.TM.down = p.adjust(enrich.TM.down, method = 'fdr')
names(enrich.TM.down) = classes
enrich.TM.down = enrich.TM.down[enrich.TM.down < 0.1]



# let's workout the results from each enrichment

enrich.BW.up      = data.frame(enrich.BW.up     ) %>% mutate(Class = rownames(.), Direction =   'up', Strain =   'BW');        
enrich.BW.down    = data.frame(enrich.BW.down   ) %>% mutate(Class = rownames(.), Direction = 'down', Strain =   'BW');       
enrich.pyrE.up    = data.frame(enrich.pyrE.up   ) %>% mutate(Class = rownames(.), Direction =   'up', Strain = 'pyrE');       
enrich.pyrE.down  = data.frame(enrich.pyrE.down ) %>% mutate(Class = rownames(.), Direction = 'down', Strain = 'pyrE');       
enrich.TM.up      = data.frame(enrich.TM.up     ) %>% mutate(Class = rownames(.), Direction =   'up', Strain =   'TM');       
enrich.TM.down    = data.frame(enrich.TM.down   ) %>% mutate(Class = rownames(.), Direction = 'down', Strain =   'TM');       


names(enrich.BW.up)[1] = 'p.value'
names(enrich.BW.down)[1] = 'p.value'
names(enrich.pyrE.up)[1] = 'p.value'
names(enrich.pyrE.down)[1] = 'p.value'
names(enrich.TM.up)[1] = 'p.value'
names(enrich.TM.down)[1] = 'p.value'


bac.enrich = rbind(enrich.BW.up, enrich.BW.down, enrich.pyrE.up, enrich.pyrE.down)

# bac.enrich %>%
#     ggplot(aes(x = Direction, y = Class)) +
#     geom_tile(aes(fill = p.value)) +
#     facet_wrap(~Strain)

# lets build a new dataframe that has all the info we need for plotting
classes = sort(unique(bac.enrich$Class))
direct = c('up', 'down')

df = expand.grid(Strain = c('BW', 'pyrE', 'TM'),
            Class = classes,
            Direction = c('up', 'down'))

df = left_join(df, bac.enrich) %>%
    mutate(p.value = replace_na(p.value, 1))

# enrichment procedure
enrbrks = c(0, -log(0.05, 10), 2, 3, 4, 100)
enrlbls = c('N.S.','<0.05','<0.01','<0.001','<0.0001')
enrcols = colorRampPalette(c("gray90", "steelblue1", "blue4"))(n = 6)

p.theme = theme(axis.ticks = element_blank(), panel.border = element_blank(), 
            panel.background = element_blank(), panel.grid.minor = element_blank(), 
            panel.grid.major = element_blank(), axis.line = element_line(colour = NA), 
            axis.line.x = element_line(colour = NA), axis.line.y = element_line(colour = NA), 
            strip.text = element_text(colour = "black", face = "bold", 
                size = 10), axis.text.x = element_text(face = "bold", 
                colour = "black", size = 10, angle = 90, hjust = 1))


# plot enrichment p-values
df %>% 
    mutate(p.value = p.value + 0.00000001,
           logFDR = ifelse(-log10(p.value) < 0, 0, -log10(p.value)),
           logFDRbin = cut(logFDR, breaks = enrbrks, labels = enrlbls, right = FALSE),
           Class = factor(Class, levels = sort(classes, decreasing = T)),
           Direction = factor(Direction, levels = sort(direct, decreasing = T))) %>%
    ggplot(aes(x = Direction, y = Class)) +
        geom_tile(aes(fill = logFDRbin)) +
        scale_fill_manual(values = enrcols) + 
        facet_wrap(~Strain) +
        # labs(fill = 'FDR') + 
        p.theme


quartz.save(file = here('Summary', 'Enrich_metab_class_bacteria.pdf'),
    type = 'pdf', dpi = 300, height = 6, width = 6)





#----------------------------#




# Enrich Worm -------------------------------------------------------------


### WORM SIDE


### initialize variables

N = length(met %in% EC_classes$EcoCycID) # total marked elements
classes = unique(EC_classes %>% filter(Plate != 'extra') %>% select(EcoCyc_Classes) %>% t %>% as.character())

join.res = res %>% 
    filter(Contrast %in% c('BW_5FU', 'TM_5FU', 'pyrE_5FU')) %>%
    left_join(worm.sum) 


# calculate enrichment values
# BW
# significative metabolites 
res.BW = join.res %>% filter(Strain == 'BW', Experiment == 'T5')
sig.met.BW = res.BW %>% filter(Metabolite != 'Negative Control', Median == 4) %>% select(EcoCycID) %>% t %>% as.vector

classes = unique(EC_classes$EcoCyc_Classes)
enrich.BW = c()
for (class in classes){
  class.met = EC_classes %>% filter(EcoCyc_Classes == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.BW)
  x = length(class.met[class.met %in% sig.met.BW])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.BW = c(enrich.BW, fit)
}

enrich.BW = p.adjust(enrich.BW, method = 'fdr')
names(enrich.BW) = classes
enrich.BW = enrich.BW[enrich.BW < 0.05]




# pyrE

res.pyrE = join.res %>% filter(Strain == 'pyrE', Experiment == 'T5')
sig.met.pyrE = res.pyrE %>% filter(Metabolite != 'Negative Control', Median == 4) %>% select(EcoCycID) %>% t %>% as.vector

classes = unique(EC_classes$EcoCyc_Classes)
enrich.pyrE = c()
for (class in classes){
  class.met = EC_classes %>% filter(EcoCyc_Classes == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.pyrE)
  x = length(class.met[class.met %in% sig.met.pyrE])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.pyrE = c(enrich.pyrE, fit)
}

enrich.pyrE = p.adjust(enrich.pyrE, method = 'fdr')
names(enrich.pyrE) = classes
enrich.pyrE = enrich.pyrE[enrich.pyrE < 0.05]

# TM
res.TM = join.res %>% filter(Strain == 'TM', Experiment == 'T250')
sig.met.TM = res.TM %>% filter(Metabolite != 'Negative Control', Median == 4) %>% select(EcoCycID) %>% t %>% as.vector


classes = unique(EC_classes$EcoCyc_Classes)
enrich.TM = c()
for (class in classes){
  class.met = EC_classes %>% filter(EcoCyc_Classes == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.TM)
  x = length(class.met[class.met %in% sig.met.TM])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.TM = c(enrich.TM, fit)
}

enrich.TM = p.adjust(enrich.TM, method = 'fdr')
names(enrich.TM) = classes
enrich.TM = enrich.TM[enrich.TM < 0.05]




# let's workout the results from each enrichment

enrich.BW      = data.frame(enrich.BW     ) %>% mutate(Class = rownames(.), Strain =   'BW');        
enrich.pyrE    = data.frame(enrich.pyrE   ) %>% mutate(Class = rownames(.), Strain = 'pyrE');       
enrich.TM      = data.frame(enrich.TM     ) %>% mutate(Class = rownames(.), Strain =   'TM');       


names(enrich.BW)[1] = 'p.value'
names(enrich.pyrE)[1] = 'p.value'
names(enrich.TM)[1] = 'p.value'

worm.enrich = rbind(enrich.BW, enrich.pyrE, enrich.TM)


# lets build a new dataframe that has all the info we need for plotting
classes = sort(unique(worm.enrich$Class))

df = expand.grid(Strain = c('BW', 'pyrE', 'TM'),
            Class = classes)

df = left_join(df, worm.enrich) %>%
    mutate(p.value = replace_na(p.value, 1))

# enrichment procedure
enrbrks = c(0, -log(0.05, 10), 2, 3, 4, 100)
enrlbls = c('N.S.','<0.05','<0.01','<0.001','<0.0001')
enrcols = colorRampPalette(c("gray90", "steelblue1", "blue4"))(n = 6)

p.theme = theme(axis.ticks = element_blank(), panel.border = element_blank(), 
            panel.background = element_blank(), panel.grid.minor = element_blank(), 
            panel.grid.major = element_blank(), axis.line = element_line(colour = NA), 
            axis.line.x = element_line(colour = NA), axis.line.y = element_line(colour = NA), 
            strip.text = element_text(colour = "black", face = "bold", 
                size = 10), axis.text.x = element_text(face = "bold", 
                colour = "black", size = 10, angle = 90, hjust = 1))



# plot enrichment p-values
df %>% 
    # arrange((Class)) %>%
    mutate(p.value = p.value + 0.00000001,
           logFDR = ifelse(-log10(p.value) < 0, 0, -log10(p.value)),
           logFDRbin = cut(logFDR, breaks = enrbrks, labels = enrlbls, right = FALSE),
           Class = factor(Class, levels = sort(classes, decreasing = T))) %>%
    ggplot(aes(x = Strain, y = Class)) +
        geom_tile(aes(fill = logFDRbin)) +
        scale_fill_manual(values = enrcols) + 
        # facet_wrap(~Strain) +
        # labs(fill = 'FDR') + 
        p.theme

quartz.save(file = here('Summary', 'Enrich_metab_class_worm.pdf'),
    type = 'pdf', dpi = 300, height = 6, width = 6)







##########################
### pathway enrichment ###
##########################


EC_path = read_csv(here('Bacteria data',"All_results_side_by_side_withKEGGEcoCycClass.csv")) %>%
                    select(Plate:EcoCyc_Classes) %>%
                    separate_rows(Pathways, sep = ';')

### calculate enrichment values
### BW
# up 
classes = unique(EC_path$Pathways)
classes = classes[-16]
classes = classes[-89]




###
# let's do it with every set

res.BW = res %>% filter(Strain == 'BW', Contrast == 'BW_5FU')
res.pyrE = res %>% filter(Strain == 'pyrE', Contrast == 'pyrE_5FU')
res.TM = res %>% filter(Strain == 'TM', Contrast == 'TM_5FU')


# total metab
met = res.BW %>% filter(Metabolite != 'Negative Control') %>% select(EcoCycID) %>% t %>% as.vector

# significative metabolites 
sig.met.BW.up =   res.BW %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC > 0) %>% select(EcoCycID) %>% t %>% as.vector
sig.met.BW.down = res.BW %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC < 0) %>% select(EcoCycID) %>% t %>% as.vector

sig.met.pyrE.up =   res.pyrE %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC > 0) %>% select(EcoCycID) %>% t %>% as.vector
sig.met.pyrE.down = res.pyrE %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC < 0) %>% select(EcoCycID) %>% t %>% as.vector

sig.met.TM.up =   res.TM %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC > 0) %>% select(EcoCycID) %>% t %>% as.vector
sig.met.TM.down = res.TM %>% filter(Metabolite != 'Negative Control', FDR < 0.05, logFC < 0) %>% select(EcoCycID) %>% t %>% as.vector


enrich.BW.up = c()
for (class in classes){
  class.met = EC_path %>% filter(Pathways == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.BW.up)
  x = length(class.met[class.met %in% sig.met.BW.up])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.BW.up = c(enrich.BW.up, fit)
}

enrich.BW.up = p.adjust(enrich.BW.up, method = 'fdr')
names(enrich.BW.up) = classes
enrich.BW.up = enrich.BW.up[enrich.BW.up < 0.05]

# down 
enrich.BW.down = c()
for (class in classes){
  class.met = EC_path %>% filter(Pathways == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.BW.down)
  x = length(class.met[class.met %in% sig.met.BW.down])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.BW.down = c(enrich.BW.down, fit)
}

enrich.BW.down = p.adjust(enrich.BW.down, method = 'fdr')
names(enrich.BW.down) = classes
enrich.BW.down = enrich.BW.down[enrich.BW.down < 0.05]





### pyrE
# up
classes = unique(EC_path$Pathways)
enrich.pyrE.up = c()
for (class in classes){
  class.met = EC_path %>% filter(Pathways == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.pyrE.up)
  x = length(class.met[class.met %in% sig.met.pyrE.up])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.pyrE.up = c(enrich.pyrE.up, fit)
}

enrich.pyrE.up = p.adjust(enrich.pyrE.up, method = 'fdr')
names(enrich.pyrE.up) = classes
enrich.pyrE.up = enrich.pyrE.up[enrich.pyrE.up < 0.05]

# down
classes = unique(EC_path$Pathways)
enrich.pyrE.down = c()
for (class in classes){
  class.met = EC_path %>% filter(Pathways == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.pyrE.down)
  x = length(class.met[class.met %in% sig.met.pyrE.down])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.pyrE.down = c(enrich.pyrE.down, fit)
}

enrich.pyrE.down = p.adjust(enrich.pyrE.down, method = 'fdr')
names(enrich.pyrE.down) = classes
enrich.pyrE.down = enrich.pyrE.down[enrich.pyrE.down < 0.05]







### TM
# up
classes = unique(EC_path$Pathways)
enrich.TM.up = c()
for (class in classes){
  class.met = EC_path %>% filter(Pathways == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.TM.up)
  x = length(class.met[class.met %in% sig.met.TM.up])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.TM.up = c(enrich.TM.up, fit)
}

enrich.TM.up = p.adjust(enrich.TM.up, method = 'fdr')
names(enrich.TM.up) = classes
enrich.TM.up = enrich.TM.up[enrich.TM.up < 0.05]


# down
classes = unique(EC_path$Pathways)
enrich.TM.down = c()
for (class in classes){
  class.met = EC_path %>% filter(Pathways == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.TM.down)
  x = length(class.met[class.met %in% sig.met.TM.down])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.TM.down = c(enrich.TM.down, fit)
}

enrich.TM.down = p.adjust(enrich.TM.down, method = 'fdr')
names(enrich.TM.down) = classes
enrich.TM.down = enrich.TM.down[enrich.TM.down < 0.05]



# let's workout the results from each enrichment

enrich.BW.up      = data.frame(enrich.BW.up     ) %>% mutate(Class = rownames(.), Direction =   'up', Strain =   'BW');        
enrich.BW.down    = data.frame(enrich.BW.down   ) %>% mutate(Class = rownames(.), Direction = 'down', Strain =   'BW');       
enrich.pyrE.up    = data.frame(enrich.pyrE.up   ) %>% mutate(Class = rownames(.), Direction =   'up', Strain = 'pyrE');       
enrich.pyrE.down  = data.frame(enrich.pyrE.down ) %>% mutate(Class = rownames(.), Direction = 'down', Strain = 'pyrE');       
enrich.TM.up      = data.frame(enrich.TM.up     ) %>% mutate(Class = rownames(.), Direction =   'up', Strain =   'TM');       
enrich.TM.down    = data.frame(enrich.TM.down   ) %>% mutate(Class = rownames(.), Direction = 'down', Strain =   'TM');       


names(enrich.BW.up)[1] = 'p.value'
names(enrich.BW.down)[1] = 'p.value'
names(enrich.pyrE.up)[1] = 'p.value'
names(enrich.pyrE.down)[1] = 'p.value'
names(enrich.TM.up)[1] = 'p.value'
names(enrich.TM.down)[1] = 'p.value'


bac.enrich = rbind(enrich.BW.up, enrich.BW.down, enrich.pyrE.up, enrich.pyrE.down)

# bac.enrich %>%
#     ggplot(aes(x = Direction, y = Class)) +
#     geom_tile(aes(fill = p.value)) +
#     facet_wrap(~Strain)

# lets build a new dataframe that has all the info we need for plotting
classes = sort(unique(bac.enrich$Class))
direct = c('up', 'down')

df = expand.grid(Strain = c('BW', 'pyrE', 'TM'),
            Class = classes,
            Direction = c('up', 'down'))

df = left_join(df, bac.enrich) %>%
    mutate(p.value = replace_na(p.value, 1))

# enrichment procedure
enrbrks = c(0, -log(0.05, 10), 2, 3, 4, 100)
enrlbls = c('N.S.','<0.05','<0.01','<0.001','<0.0001')
enrcols = colorRampPalette(c("gray90", "steelblue1", "blue4"))(n = 6)

p.theme = theme(axis.ticks = element_blank(), panel.border = element_blank(), 
            panel.background = element_blank(), panel.grid.minor = element_blank(), 
            panel.grid.major = element_blank(), axis.line = element_line(colour = NA), 
            axis.line.x = element_line(colour = NA), axis.line.y = element_line(colour = NA), 
            strip.text = element_text(colour = "black", face = "bold", 
                size = 10), axis.text.x = element_text(face = "bold", 
                colour = "black", size = 10, angle = 90, hjust = 1))


# plot enrichment p-values
df %>% 
    mutate(p.value = p.value + 0.00000001,
           logFDR = ifelse(-log10(p.value) < 0, 0, -log10(p.value)),
           logFDRbin = cut(logFDR, breaks = enrbrks, labels = enrlbls, right = FALSE),
           Class = factor(Class, levels = sort(classes, decreasing = T)),
           Direction = factor(Direction, levels = sort(direct, decreasing = T))) %>%
    ggplot(aes(x = Direction, y = Class)) +
        geom_tile(aes(fill = logFDRbin)) +
        scale_fill_manual(values = enrcols) + 
        facet_wrap(~Strain) +
        # labs(fill = 'FDR') + 
        p.theme


quartz.save(file = here('Summary', 'Enrich_pathways_bacteria.pdf'),
    type = 'pdf', dpi = 300, height = 6, width = 6)







### WORM SIDE

join.res = res %>% 
    filter(Contrast %in% c('BW_5FU', 'TM_5FU', 'pyrE_5FU')) %>%
    left_join(worm.sum) 


# calculate enrichment values
# BW
# significative metabolites 
res.BW = join.res %>% filter(Strain == 'BW', Experiment == 'T5')
sig.met.BW = res.BW %>% filter(Metabolite != 'Negative Control', Median == 4) %>% select(EcoCycID) %>% t %>% as.vector

classes = unique(EC_path$Pathways)

enrich.BW = c()
for (class in classes){
  class.met = EC_path %>% filter(Pathways == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.BW)
  x = length(class.met[class.met %in% sig.met.BW])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.BW = c(enrich.BW, fit)
}

enrich.BW = p.adjust(enrich.BW, method = 'fdr')
names(enrich.BW) = classes
enrich.BW = enrich.BW[enrich.BW < 0.05]




# pyrE

res.pyrE = join.res %>% filter(Strain == 'pyrE', Experiment == 'T5')
sig.met.pyrE = res.pyrE %>% filter(Metabolite != 'Negative Control', Median == 4) %>% select(EcoCycID) %>% t %>% as.vector

enrich.pyrE = c()
for (class in classes){
  class.met = EC_path %>% filter(Pathways == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.pyrE)
  x = length(class.met[class.met %in% sig.met.pyrE])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.pyrE = c(enrich.pyrE, fit)
}

enrich.pyrE = p.adjust(enrich.pyrE, method = 'fdr')
names(enrich.pyrE) = classes
enrich.pyrE = enrich.pyrE[enrich.pyrE < 0.05]

# TM
res.TM = join.res %>% filter(Strain == 'TM', Experiment == 'T250')
sig.met.TM = res.TM %>% filter(Metabolite != 'Negative Control', Median == 4) %>% select(EcoCycID) %>% t %>% as.vector


enrich.TM = c()
for (class in classes){
  class.met = EC_path %>% filter(Pathways == class) %>% select(EcoCycID) %>% t %>% as.vector
  n = N - m
  k = length(sig.met.TM)
  x = length(class.met[class.met %in% sig.met.TM])
  fit = phyper(q = x-1, m = m, n = n, k = k, lower.tail = FALSE)
  enrich.TM = c(enrich.TM, fit)
}

enrich.TM = p.adjust(enrich.TM, method = 'fdr')
names(enrich.TM) = classes
enrich.TM = enrich.TM[enrich.TM < 0.05]




# let's workout the results from each enrichment

enrich.BW      = data.frame(enrich.BW     ) %>% mutate(Class = rownames(.), Strain =   'BW');        
enrich.pyrE    = data.frame(enrich.pyrE   ) %>% mutate(Class = rownames(.), Strain = 'pyrE');       
enrich.TM      = data.frame(enrich.TM     ) %>% mutate(Class = rownames(.), Strain =   'TM');       


names(enrich.BW)[1] = 'p.value'
names(enrich.pyrE)[1] = 'p.value'
names(enrich.TM)[1] = 'p.value'

worm.enrich = rbind(enrich.BW, enrich.pyrE, enrich.TM)


# lets build a new dataframe that has all the info we need for plotting
classes = sort(unique(worm.enrich$Class))

df = expand.grid(Strain = c('BW', 'pyrE', 'TM'),
            Class = classes)

df = left_join(df, worm.enrich) %>%
    mutate(p.value = replace_na(p.value, 1))

# enrichment procedure
enrbrks = c(0, -log(0.05, 10), 2, 3, 4, 100)
enrlbls = c('N.S.','<0.05','<0.01','<0.001','<0.0001')
enrcols = colorRampPalette(c("gray90", "steelblue1", "blue4"))(n = 6)

p.theme = theme(axis.ticks = element_blank(), panel.border = element_blank(), 
            panel.background = element_blank(), panel.grid.minor = element_blank(), 
            panel.grid.major = element_blank(), axis.line = element_line(colour = NA), 
            axis.line.x = element_line(colour = NA), axis.line.y = element_line(colour = NA), 
            strip.text = element_text(colour = "black", face = "bold", 
                size = 10), axis.text.x = element_text(face = "bold", 
                colour = "black", size = 10, angle = 90, hjust = 1))



# plot enrichment p-values
df %>% 
    # arrange((Class)) %>%
    mutate(p.value = p.value + 0.00000001,
           logFDR = ifelse(-log10(p.value) < 0, 0, -log10(p.value)),
           logFDRbin = cut(logFDR, breaks = enrbrks, labels = enrlbls, right = FALSE),
           Class = factor(Class, levels = sort(classes, decreasing = T))) %>%
    ggplot(aes(x = Strain, y = Class)) +
        geom_tile(aes(fill = logFDRbin)) +
        scale_fill_manual(values = enrcols) + 
        # facet_wrap(~Strain) +
        # labs(fill = 'FDR') + 
        p.theme

quartz.save(file = here('Summary', 'Enrich_pathways_worm.pdf'),
    type = 'pdf', dpi = 300, height = 6, width = 6)























###########################
### sub-library screens ###
###########################

##########################
### Galactose and Glycerol
##########################

# read and transform data
galactose = read_xlsx('Sub_library/Summary_Keio_sublibrary_Galactose_Glycerol.xlsx', sheet = 'Summary_good') 

galactose = galactose %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% # make table long
    unite(ID, Genes, Supplement, Replicate, remove = FALSE) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Genes = as.factor(Genes),
        Pathway = as.factor(Pathway),
        ID = as.factor(ID),
        Drug = as.factor(Drug)) %>%
    select(Genes, Well, ID, Pathway, Replicate, Supplement, Supplement_mM, Drug, Score)


# table with genotypes and pathways

pathways = galactose %>% select(Genes, Pathway) %>% distinct(GGenesenotype, .keep_all = TRUE)
pathways = unique(pathways)
# data summary
gal.sum = galactose %>%
  group_by(Supplement, Supplement_mM, Drug, Genes) %>%
  summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
  mutate(BW_Score = Median_Score[Genes == 'BW'],
         BW_Mean = Mean[Genes == 'BW']) %>%
  ungroup %>%
  group_by(Genes, Drug) %>%
  ungroup %>%
  mutate(BW_norm = Median_Score - BW_Score,
         BW_Mean_norm = Mean - BW_Mean)


### SCATTER PLOT FROM GLYCEROL

pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
gal.sum %>% 
    filter(Supplement == 'Glycerol', Drug == 5) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    left_join(pathways) %>% 
    ggplot(aes(x = Glycerol_0, y = Glycerol_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(aes(colour = Pathway),position = pos, size = 2) + 
    # scale_color_manual(values = grad) + 
    # scale_fill_manual(values = grad) +
    # geom_text_repel(aes(label = ifelse(Genes == 'ppnP', as.character(Genes), '')), position = pos) + # this point went rogue
    geom_text_repel(aes(label = Genes), position = pos) +
    labs(title = expression(paste("5FU + Glycerol effect on ", italic('C. elegans'), " phenotype", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Glycerol)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', 'Scatter_sub_lib_glycerol_5uM.pdf'),
    type = 'pdf', dpi = 300, height = 10, width = 12)



### SCATTER PLOT FROM GALACTOSE

pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
gal.sum %>% 
    filter(Supplement == 'Galactose', Drug == 5) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    left_join(pathways) %>% 
    ggplot(aes(x = Galactose_0, y = Galactose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(aes(colour = Pathway),position = pos, size = 2) + 
    # scale_color_manual(values = grad) + 
    # scale_fill_manual(values = grad) +
    # geom_text_repel(aes(label = ifelse(Genes == 'ppnP', as.character(Genes), '')), position = pos) + # this point went rogue
    geom_text_repel(aes(label = Genes), position = pos) +
    labs(title = expression(paste("5FU + Galactose effect on ", italic('C. elegans'), " phenotype", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Galactose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', 'Scatter_sub_lib_galactose_5uM.pdf'),
    type = 'pdf', dpi = 300, height = 10, width = 12)


### BARPLOTS

# version 2
# set variable names for the facet wrap
# this is a general bargraph to plot everything 
variable_names = c(
  '1' = "1 mM Glycerol",
  '10' = "10 mM Glycerol"
)


# glycerol
gal.sum %>%
    ungroup %>%
    filter(Supplement == 'Glycerol', !Drug == 2.5, !Genes == 'thyA') %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = galactose %>% 
    						filter(Supplement == 'Glycerol', !Drug == 2.5, !Genes == 'thyA') %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A')) +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    geom_vline(xintercept = 1.5, size = 0.8) +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'barplot_all_glycerol.pdf'),
    type = 'pdf', dpi = 300, height = 9, width = 14)


# galactose
gal.sum %>%
    ungroup %>%
    filter(Supplement == 'Galactose', !Drug == 2.5, !Genes == 'thyA') %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = galactose %>% 
    						filter(Supplement == 'Galactose', !Drug == 2.5, !Genes == 'thyA') %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    geom_vline(xintercept = 1.5, size = 0.8) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A')) +
    # scale_fill_discrete_sequential(palette = "Reds2", nmax = 6, order = 3:6) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'barplot_all_galactose.pdf'),
    type = 'pdf', dpi = 300, height = 9, width = 14)




### subplots by drug concentration


# galactose
drug = 5
supp = 'Galactose'
gal.sum %>%
    ungroup %>%
    filter(Supplement == supp, Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = galactose %>% 
    						filter(Supplement == supp, Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
    	axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_sub_lib_',supp,'_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 8, width = 11)






### Median/lm test

### user functions 
# function to 'tidy' the output of the median function 
tidy.med = function(model) {
    df = data.frame(model[1], model[2], model[3], model[4], model[5])
    df = as_tibble(df)
    return(df)
}


# function that calculates the comparison between BW and other genes in the datasets
# median version!
m.test = function(data, genes) {
    results = data.frame()
    for (i in 1:length(genes)){
        sub = data %>% filter(Genes %in% c('BW', genes[i]))
        if(dim(sub)[1] <= 2) {next} # skips genes that are not present (kinda)
        model = mood.medtest(Sum ~ Genes, data = sub, exact = FALSE)
        df = tidy.med(model)
        df['Gene'] = genes[i]
        df['Supplement'] = unique(data$Supplement)
        df['Supplement_mM'] = unique(data$Supplement_mM)
        results = rbind(results,df)
    }
    results = results %>% mutate(p.stars = stars.pval(p.value), 
            fdr = p.adjust(p.value, method = 'fdr'), fdr.stars = stars.pval(fdr)) %>%
        select(Gene, Supplement, Supplement_mM,method,  statistic, parameter, p.value, p.stars,   fdr, fdr.stars)
    return(results)
}

# function that calculates the comparison between BW and other genes in the datasets
# lm version!
av.test = function(data, genes) {
    results = data.frame()
    for (i in 1:length(genes)){
        sub = data %>% filter(Genes %in% c('BW', genes[i]))
        if(dim(sub)[1] <= 2) {next} # skips genes that are not present (kinda)
        model = lm(Sum ~ Genes, data = sub)
        df = tidy(model)[2,]
        df['Gene'] = genes[i]
        df['Supplement'] = unique(data$Supplement)
        df['Supplement_mM'] = unique(data$Supplement_mM)
        results = rbind(results,df)
    }
    results = results %>% mutate(p.stars = stars.pval(p.value), 
            fdr = p.adjust(p.value, method = 'fdr'), fdr.stars = stars.pval(fdr)) %>%
        select(Gene, Supplement, Supplement_mM, estimate, std.error, statistic, p.value, p.stars, fdr, fdr.stars)
    return(results)
}


# stupid function to remove the duplicated BWs (it does the average of them per replicate)
ctrls = function(data){
    ctr = data %>% filter(Genes == 'BW') %>% 
        group_by(Genes, Replicate, Supplement, Supplement_mM) %>% 
        summarise(Sum = median(Sum, na.rm = TRUE)) %>% ungroup
    # exclude the 'old' BW and include the average 
    data = data %>% mutate(Genes = as.character(Genes)) %>% filter(!Genes == 'BW') 
    data = rbind(ctr, data)
    return(data)
}


###--------------------------


# first, sum all score values by gene and condition

gal.auc = galactose %>% group_by(Genes, Well, Replicate, Supplement, Supplement_mM) %>%
    filter(!Drug == 2.5) %>%
    summarise(Sum = sum(Score)) 

# filter values with NAs (maybe try to impute them with RF?)
gal.auc = gal.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 




## we want to compare each gene against BW per condition
# split data
gal_0 = gal.auc %>% filter(Supplement == 'Galactose', Supplement_mM == 0) %>% ungroup %>% select(-Well)
gal_10 = gal.auc %>% filter(Supplement == 'Galactose', Supplement_mM == 10) %>% ungroup %>% select(-Well)
gly_0 = gal.auc %>% filter(Supplement == 'Glycerol', Supplement_mM == 0) %>% ungroup %>% select(-Well)
gly_10 = gal.auc %>% filter(Supplement == 'Glycerol', Supplement_mM == 10) %>% ungroup %>% select(-Well)

# remove BWs with my dummy function
gal_0 = ctrls(gal_0)
gal_10 = ctrls(gal_10)
gly_0 = ctrls(gly_0)
gly_10 = ctrls(gly_10)

# extract the gene list for the loop
genes = galactose %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# calculate the Mood's test (median test) for every condition
med.res = rbind(m.test(gal_0, genes), m.test(gal_10, genes), m.test(gly_0, genes), m.test(gly_10, genes))
# the same but for the t.test
ttest.res = rbind(av.test(gal_0, genes), av.test(gal_10, genes), av.test(gly_0, genes), av.test(gly_10, genes))


## after having done the statistical tests, we can filter those ones that have sig. differences
# for this dataset I will use the t test results, as the other is completely useless (data too variable, and only 2 reps)

gal.genes = ttest.res %>% filter(Supplement == 'Galactose',p.value <= 0.05) %>% select(Gene) %>% unique
gly.genes = ttest.res %>% filter(Supplement == 'Glycerol',p.value <= 0.05) %>% select(Gene) %>% unique

gal.genes = sort(as.character(gal.genes$Gene)) ; gal.genes = c('BW', gal.genes)
gly.genes = sort(as.character(gly.genes$Gene)) ; gly.genes = c('BW', gly.genes)


gal.genes = factor(gal.genes, levels = gal.genes)
gly.genes = factor(gly.genes, levels = gly.genes)



# AND NOW WE CAN PLOT IT!


# glycerol
gal.sum %>%
    ungroup %>%
    filter(Genes %in% gal.genes, Supplement == 'Glycerol', !Drug == 2.5) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = galactose %>% 
                            filter(Genes %in% gal.genes, Supplement == 'Glycerol', !Drug == 2.5) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A')) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'Bargraph_hits_sub_lib_glycerol.pdf'),
    type = 'pdf', dpi = 300, height = 7, width = 9)


# galactose
gal.sum %>%
    ungroup %>%
    filter(Genes %in% gly.genes, Supplement == 'Galactose', !Drug == 2.5) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = galactose %>% 
                            filter(Genes %in% gly.genes, Supplement == 'Galactose', !Drug == 2.5) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    geom_vline(xintercept = 1.5, size = 0.8) +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A')) +
    # scale_fill_discrete_sequential(palette = "Reds2", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'Bargraph_hits_sub_lib_galactose.pdf'),
    type = 'pdf', dpi = 300, height = 7, width = 9)



# save statistics list
list_of_datasets = list('Summary statistics' = gal.sum, 'AUC' = gal.auc , 't-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'galactose_glycerol_stats.xlsx'), colNames = T, rowNames = F) 




# barplots for AUC comparison

length(gal.genes)

# generate a small dataframe to divide the original dataframe
div = c(rep(1, 10), rep(2, 9))
df = data.frame(Genes = gal.genes, div)


sup = 'Glycerol'
genes = gly.genes
gal.auc %>%
    ungroup %>%
    filter(Genes %in% genes, Supplement == sup) %>%
    mutate(Genes = factor(Genes, levels = genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    # ungroup %>%
    # mutate(Genes = as.character(Genes), Supplement_mM = as.character(Supplement_mM)) %>%
    ggplot(aes(x = Genes, y = Mean, fill = Supplement_mM,  width = 0.9)) +
    geom_bar(aes(x = Genes), stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = gal.auc %>% 
        filter(Genes %in% genes, Supplement == sup) %>% ungroup %>%
        left_join(df), aes(x = Genes, y = Sum), 
                            position = position_jitterdodge(jitter.width = 0.4, jitter.height = 0.05, dodge.width = 1), 
                            alpha = 0.8, show.legend = FALSE) +
    theme_light() +
        labs(x = 'Genes',
             y = 'AUC mean') +
    # facet_wrap(~div, scales = "free_x", nrow = 2) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', paste0('Bargraph_AUC_hits_sub_lib_', sup,'.pdf')),
    type = 'pdf', dpi = 300, height = 6, width = 10)


sup = 'Galactose'
genes = gal.genes
gal.auc %>%
    ungroup %>%
    filter(Genes %in% genes, Supplement == sup) %>%
    mutate(Genes = factor(Genes, levels = genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Supplement_mM)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = gal.auc %>% 
        filter(Genes %in% genes, Supplement == sup) %>% ungroup %>%
        mutate(Genes = factor(Genes, levels = genes)) %>%
        left_join(df), aes(x = Genes, y = Sum, group = Supplement_mM), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    theme(strip.text = element_text(colour = 'black'),
        # axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', paste0('Pointrange_AUC_hits_sub_lib_', sup,'.pdf')),
    type = 'pdf', dpi = 300, height = 4, width = 9)




#---------------------------------------



##########################
### Glucose sublibrary ###
##########################

# read and transform data
glucose = read_xlsx('Sub_library/Summary_Keio Sublibrary_Glucose Supp_10mM_N2 worms.xlsx', sheet = 'Summary_good') 

glucose = glucose %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% # make table long
    unite(ID, Genes, Supplement, Replicate, remove = FALSE) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Genes = as.factor(Genes),
        Pathway = as.factor(Pathway),
        ID = as.factor(ID),
        Drug = as.factor(Drug)) %>%
    select(Genes, Well, ID, Pathway, Replicate, Supplement, Supplement_mM, Drug, Score)


# table with genotypes and pathways

pathways = glucose %>% select(Genes, Pathway) %>% unique

# data summary
glu.sum = glucose %>%
  group_by(Supplement, Supplement_mM, Drug, Genes) %>%
  summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
  mutate(BW_Score = Median_Score[Genes == 'BW'],
         BW_Mean = Mean[Genes == 'BW']) %>%
  ungroup %>%
  group_by(Genes, Drug) %>%
  ungroup %>%
  mutate(BW_norm = Median_Score - BW_Score,
         BW_Mean_norm = Mean - BW_Mean)


### SCATTER PLOT

drug = 5
pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
glu.sum %>% 
    filter(Drug == drug) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    left_join(pathways) %>% 
    ggplot(aes(x = Glucose_0, y = Glucose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(aes(colour = Pathway),position = pos, size = 2) + 
    # scale_color_manual(values = grad) + 
    # scale_fill_manual(values = grad) +
    # geom_text_repel(aes(label = ifelse(Genes == 'ppnP', as.character(Genes), '')), position = pos) + # this point went rogue
    geom_text_repel(aes(label = Genes), position = pos, max.overlaps = 100) +
    labs(title = expression(paste("5FU + Glucose effect on ", italic('C. elegans'), " N2 phenotype", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' N2 phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' N2 phenotype ' , bold('(Glucose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', paste0('Scatter_sub_lib_glucose_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 10, width = 12)



### Bargraphs

# all glucose barplot
glu.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = glucose %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8, 
                            show.legend = FALSE) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'barplot_all_sub_lib_glucose.pdf'),
    type = 'pdf', dpi = 300, height = 9, width = 14)



# barplot per drug concentration
drug = 5
glu.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = glucose %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_sub_lib_glucose_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 8, width = 11)





###--------------------------


# first, sum all score values by gene and condition
glu.auc = glucose %>% group_by(Genes, Well, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score)) 

# filter values with NAs (maybe try to impute them with RF?)
glu.auc = glu.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


## we want to compare each gene against BW per condition
# split data
glu_0 = glu.auc %>%  filter(Supplement_mM == 0) %>% ungroup %>% select(-Well)
glu_10 = glu.auc %>% filter(Supplement_mM == 10) %>% ungroup %>% select(-Well)


# remove BWs with my dummy function
glu_0 = ctrls(glu_0)
glu_10 = ctrls(glu_10)

# extract the gene list for the loop
genes = glucose %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# calculate the Mood's test (median test) for every condition
med.res = rbind(m.test(glu_0, genes), m.test(glu_10, genes))
# the same but for the t.test
ttest.res = rbind(av.test(glu_0, genes), av.test(glu_10, genes))


## after having done the statistical tests, we can filter those ones that have sig. differences
# for this dataset I will use the t test results, as the other is completely useless (data too variable, and only 2 reps)

glu.genes = ttest.res %>% filter(fdr <= 0.05) %>% select(Gene) %>% unique
glu.genes = sort(as.character(glu.genes$Gene)) ; glu.genes = c('BW', glu.genes)
glu.genes = factor(glu.genes, levels = glu.genes)

## plot hits for glucose

glu.sum %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Drug = as.factor(Drug),
        Genes = factor(Genes, levels = glu.genes)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black', show.legend = FALSE) +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = glucose %>% 
        filter(Genes %in% glu.genes) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8,show.legend = FALSE) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A')) +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'Bargraph_hits_sub_lib_glucose.pdf'),
    type = 'pdf', dpi = 300, height = 13, width = 17)




# barplots for AUC comparison

length(glu.genes)

# generate a small dataframe to divide the original dataframe
div = c(rep(1, 16), rep(2, 15), rep(3, 15))
df = data.frame(Genes = glu.genes, div)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, fill = Supplement_mM,  width = 0.9)) +
    geom_bar(aes(x = Genes), stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = as.character(Genes), Supplement_mM = as.character(Supplement_mM)) %>%
        left_join(df), aes(x = Genes, y = Sum), 
                            position = position_jitterdodge(jitter.width = 0.4, jitter.height = 0.05, dodge.width = 1), 
                            alpha = 0.8, show.legend = FALSE) +
    theme_light() +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 3) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Bargraph_AUC_hits_sub_lib_glucose.pdf'),
    type = 'pdf', dpi = 300, height = 13, width = 14)






# version with pointrange (not very polished at the moment)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Supplement_mM)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = factor(Genes, levels = glu.genes)) %>%
        left_join(df), aes(x = Genes, y = Sum, group = Supplement_mM), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 3) +
    theme(strip.text = element_text(colour = 'black'),
        # axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Pointrange_AUC_hits_sub_lib_glucose.pdf'),
    type = 'pdf', dpi = 300, height = 8, width = 9)



# experimental plot

# # prepare df for means
# temp = auc.sum %>% ungroup %>% 
#     select(Genes, Supplement_mM, Mean) %>%
#     mutate(Var = 'Mean') %>%
#     unite(ID, Var, Supplement_mM) %>%
#     spread(ID, Mean)

# # prepare df for sd
# temp2 = auc.sum %>% ungroup %>% 
#     select(Genes, Supplement_mM, SD) %>%
#     mutate(Var = 'SD') %>%
#     unite(ID, Var, Supplement_mM) %>%
#     spread(ID, SD)

# left_join(temp, temp2) %>%
#     left_join(pathways) %>%
#     ggplot(aes(x = Mean_0, y = Mean_10)) +
#     geom_point(size = 4, alpha = 1) +
#     geom_errorbar(aes(ymin = Mean_10 - SD_10, ymax = Mean_10 + SD_10), width = 0.1) +
#     geom_errorbarh(aes(xmin = Mean_0 - SD_0, xmax = Mean_0 + SD_0), height = 0.1) +
#     geom_text_repel(aes(label = Genes)) +
#     theme_light() +
#         labs(x = 'AUC mean',
#              y = 'AUC mean (with supplement)')


# left_join(temp, temp2) %>%
#     left_join(pathways) %>%
#     ggplot(aes(x = Mean_0, y = Mean_10)) +
#     # geom_point(size = 4, alpha = 1) +
#     geom_errorbar(aes(ymin = Mean_10 - SD_10, ymax = Mean_10 + SD_10), width = 0) +
#     geom_errorbarh(aes(xmin = Mean_0 - SD_0, xmax = Mean_0 + SD_0), height = 0) +
#     geom_label(aes(label = Genes)) +
#     theme_light() +
#         labs(x = 'AUC mean',
#              y = 'AUC mean (with supplement)')


# left_join(temp, temp2) %>%
#     left_join(pathways) %>%
#     ggplot(aes(x = Mean_0, y = Mean_10)) +
#     geom_point(aes(size = (SD_0 * SD_10)), alpha = 1, scale_size(range = c(0.5, 12)) +
#     # geom_errorbar(aes(ymin = Mean_10 - SD_10, ymax = Mean_10 + SD_10), width = 0) +
#     # geom_errorbarh(aes(xmin = Mean_0 - SD_0, xmax = Mean_0 + SD_0), height = 0) +
#     # geom_label(aes(label = Genes)) +
#     theme_light() +
#         labs(x = 'AUC mean',
#              y = 'AUC mean (with supplement)')






# save statistics list
list_of_datasets = list('Summary statistics' = glu.sum, 'AUC' = glu.auc, 't-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'glucose_N2_stats.xlsx'), colNames = T, rowNames = F) 







#---------------------------------------






################################
### Glucose sublibrary - pyr ###
################################


# read and transform data
glucose.pyr = read_xlsx('Sub_library/Summary_Keio Sublibrary_Glucose Supp_10mM_pyr-1 worms.xlsx', sheet = 'Summary_good') 

glucose.pyr = glucose.pyr %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% # make table long
    unite(ID, Genes, Supplement, Replicate, remove = FALSE) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Genes = as.factor(Genes),
        Pathway = as.factor(Pathway),
        ID = as.factor(ID),
        Drug = as.factor(Drug)) %>%
    select(Genes, Well, ID, Pathway, Replicate, Supplement, Supplement_mM, Drug, Score)


# table with genotypes and pathways

pathways = glucose.pyr %>% select(Genes, Pathway) %>% unique

# data summary
glu.pyr.sum = glucose.pyr %>%
  group_by(Supplement, Supplement_mM, Drug, Genes) %>%
  summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
  mutate(BW_Score = Median_Score[Genes == 'BW'],
         BW_Mean = Mean[Genes == 'BW']) %>%
  ungroup %>%
  group_by(Genes, Drug) %>%
  ungroup %>%
  mutate(BW_norm = Median_Score - BW_Score,
         BW_Mean_norm = Mean - BW_Mean)


### SCATTER PLOT

drug = 5
pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
glu.pyr.sum %>% 
    filter(Drug == drug) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    left_join(pathways) %>% 
    ggplot(aes(x = Glucose_0, y = Glucose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(aes(colour = Pathway),position = pos, size = 2) + 
    # geom_text_repel(aes(label = ifelse(Genes %in% c('ppnP', 'gltA', 'fbp', 'pgm'), as.character(Genes), '')), position = pos) + # this point went rogue
    geom_text_repel(aes(label = Genes), position = pos) +
    # coord_cartesian(ylim = c(-1, 1)) + # use in drug == 0
    labs(title = expression(paste("5FU + Glucose effect on ", italic('C. elegans'), " pyr1 phenotype", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Glucose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', paste0('Scatter_sub_lib_glucose_',as.character(drug),'uM_pyr1.pdf')),
    type = 'pdf', dpi = 300, height = 10, width = 12)



### Bargraphs

# all glucose barplot
glu.pyr.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = glucose.pyr %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    theme_light() +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'barplot_all_sub_lib_glucose_pyr1.pdf'),
    type = 'pdf', dpi = 300, height = 9, width = 14)



# barplot per drug concentration
drug = 5
glu.pyr.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glucose.pyr %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_sub_lib_glucose',as.character(drug),'uM_pyr1.pdf')),
    type = 'pdf', dpi = 300, height = 8, width = 11)





###--------------------------


# first, sum all score values by gene and condition
glu.auc = glucose.pyr %>% group_by(Genes, Well, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score)) 

# filter values with NAs (maybe try to impute them with RF?)
glu.auc = glu.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


## we want to compare each gene against BW per condition
# split data
glu_0 = glu.auc %>%  filter(Supplement_mM == 0) %>% ungroup %>% select(-Well)
glu_10 = glu.auc %>% filter(Supplement_mM == 10) %>% ungroup %>% select(-Well)


# remove BWs with my dummy function
glu_0 = ctrls(glu_0)
glu_10 = ctrls(glu_10)

# extract the gene list for the loop
genes = glucose.pyr %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# calculate the Mood's test (median test) for every condition
med.res = rbind(m.test(glu_0, genes), m.test(glu_10, genes))
# the same but for the t.test
ttest.res = rbind(av.test(glu_0, genes), av.test(glu_10, genes))


## after having done the statistical tests, we can filter those ones that have sig. differences
# for this dataset I will use the t test results, as the other is completely useless (data too variable, and only 2 reps)

glu.genes = ttest.res %>% filter(fdr <= 0.05) %>% select(Gene) %>% unique
glu.genes = sort(as.character(glu.genes$Gene)) ; glu.genes = c('BW', glu.genes)
glu.genes = factor(glu.genes, levels = glu.genes)

## plot hits for glucose

glu.pyr.sum %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = glucose.pyr %>% 
        filter(Genes %in% glu.genes) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))


quartz.save(file = here('Summary', 'Bargraph_hits_sub_lib_glucose_pyr1.pdf'),
    type = 'pdf', dpi = 300, height = 13, width = 17)



# save statistics list
list_of_datasets = list('Summary statistics' = glu.pyr.sum, 'AUC' = glu.auc, 't-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'glucose_pyr1_stats.xlsx'), colNames = T, rowNames = F) 




length(glu.genes)

# generate a small dataframe to divide the original dataframe
div = c(rep(1, 15), rep(2, 15))
df = data.frame(Genes = glu.genes, div)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, fill = Supplement_mM,  width = 0.9)) +
    geom_bar(aes(x = Genes), stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = as.character(Genes), Supplement_mM = as.character(Supplement_mM)) %>%
        left_join(df), aes(x = Genes, y = Sum), 
                            position = position_jitterdodge(jitter.width = 0.4, jitter.height = 0.05, dodge.width = 1), 
                            alpha = 0.8, show.legend = FALSE) +
    theme_light() +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 3) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Bargraph_AUC_hits_sub_lib_glucose_pyr1.pdf'),
    type = 'pdf', dpi = 300, height = 13, width = 14)



# version with pointrange (not very polished at the moment)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Supplement_mM)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = factor(Genes, levels = glu.genes)) %>%
        left_join(df), aes(x = Genes, y = Sum, group = Supplement_mM), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 3) +
    theme(strip.text = element_text(colour = 'black'),
        # axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Pointrange_AUC_hits_sub_lib_glucose_pyr1.pdf'),
    type = 'pdf', dpi = 300, height = 8, width = 9)












#################################
### Glucose sublibrary - umps ###
#################################


# read and transform data
glucose.umps = read_xlsx('Sub_library/Summary_Keio Sublibrary_Glucose Supp_10mM_umps-1 worms.xlsx', sheet = 'Summary_good') 

glucose.umps = glucose.umps %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% # make table long
    unite(ID, Genes, Supplement, Replicate, remove = FALSE) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Genes = as.factor(Genes),
        Pathway = as.factor(Pathway),
        ID = as.factor(ID),
        Drug = as.factor(Drug)) %>%
    select(Genes, Well, ID, Pathway, Replicate, Supplement, Supplement_mM, Drug, Score)


# table with genotypes and pathways

pathways = glucose.umps %>% select(Genes, Pathway) %>% unique

# data summary
glu.umps.sum = glucose.umps %>%
  group_by(Supplement, Supplement_mM, Drug, Genes) %>%
  summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
  mutate(BW_Score = Median_Score[Genes == 'BW'],
         BW_Mean = Mean[Genes == 'BW']) %>%
  ungroup %>%
  group_by(Genes, Drug) %>%
  ungroup %>%
  mutate(BW_norm = Median_Score - BW_Score,
         BW_Mean_norm = Mean - BW_Mean)


### SCATTER PLOT

drug = 5
pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
glu.umps.sum %>% 
    filter(Drug == drug) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    left_join(pathways) %>% 
    ggplot(aes(x = Glucose_0, y = Glucose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(aes(colour = Pathway),position = pos, size = 2) + 
    # geom_text_repel(aes(label = ifelse(Genes %in% c('ppnP', 'gltA', 'fbp', 'pgm'), as.character(Genes), '')), position = pos) + # this point went rogue
    geom_text_repel(aes(label = Genes), position = pos) +
    labs(title = expression(paste("5FU + Glucose effect on ", italic('C. elegans'), " umps1 phenotype", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Glucose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', paste0('Scatter_sub_lib_glucose_',as.character(drug),'uM_umps1.pdf')),
    type = 'pdf', dpi = 300, height = 10, width = 12)



### Bargraphs

# all glucose barplot
glu.umps.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = glucose.umps %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    theme_light() +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'barplot_all_sub_lib_glucose_umps1.pdf'),
    type = 'pdf', dpi = 300, height = 9, width = 14)



# barplot per drug concentration
drug = 5
glu.umps.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glucose.umps %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_sub_lib_glucose',as.character(drug),'uM_umps1.pdf')),
    type = 'pdf', dpi = 300, height = 8, width = 11)





###--------------------------


# first, sum all score values by gene and condition
glu.auc = glucose.umps %>% group_by(Genes, Well, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score)) 

# filter values with NAs (maybe try to impute them with RF?)
glu.auc = glu.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


## we want to compare each gene against BW per condition
# split data
glu_0 = glu.auc %>%  filter(Supplement_mM == 0) %>% ungroup %>% select(-Well)
glu_10 = glu.auc %>% filter(Supplement_mM == 10) %>% ungroup %>% select(-Well)


# remove BWs with my dummy function
glu_0 = ctrls(glu_0)
glu_10 = ctrls(glu_10)

# extract the gene list for the loop
genes = glucose.umps %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# calculate the Mood's test (median test) for every condition
med.res = rbind(m.test(glu_0, genes), m.test(glu_10, genes))
# the same but for the t.test
ttest.res = rbind(av.test(glu_0, genes), av.test(glu_10, genes))


## after having done the statistical tests, we can filter those ones that have sig. differences
# for this dataset I will use the t test results, as the other is completely useless (data too variable, and only 2 reps)

glu.genes = ttest.res %>% filter(p.value <= 0.05) %>% select(Gene) %>% unique
glu.genes = sort(as.character(glu.genes$Gene)) ; glu.genes = c('BW', glu.genes)
glu.genes = factor(glu.genes, levels = glu.genes)

## plot hits for glucose

glu.umps.sum %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = glucose.umps %>% 
        filter(Genes %in% glu.genes) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))


quartz.save(file = here('Summary', 'Bargraph_hits_sub_lib_glucose_umps1.pdf'),
    type = 'pdf', dpi = 300, height = 13, width = 17)



# save statistics list
list_of_datasets = list('Summary statistics' = glu.umps.sum, 'AUC' = glu.auc, 't-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'glucose_umps1_stats.xlsx'), colNames = T, rowNames = F) 




length(glu.genes)

# generate a small dataframe to divide the original dataframe
div = c(rep(1, 10), rep(2, 10))
df = data.frame(Genes = glu.genes, div)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, fill = Supplement_mM,  width = 0.9)) +
    geom_bar(aes(x = Genes), stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = as.character(Genes), Supplement_mM = as.character(Supplement_mM)) %>%
        left_join(df), aes(x = Genes, y = Sum), 
                            position = position_jitterdodge(jitter.width = 0.4, jitter.height = 0.05, dodge.width = 1), 
                            alpha = 0.8, show.legend = FALSE) +
    theme_light() +
        labs(x = 'Genes',
             y = 'AUC mean') +
    # facet_wrap(~div, scales = "free_x", nrow = 3) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Bargraph_AUC_hits_sub_lib_glucose_umps1.pdf'),
    type = 'pdf', dpi = 300, height = 8, width = 11)



# version with pointrange 

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Supplement_mM)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = factor(Genes, levels = glu.genes)) %>%
        left_join(df), aes(x = Genes, y = Sum, group = Supplement_mM), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    # facet_wrap(~div, scales = "free_x", nrow = 3) +
    theme(strip.text = element_text(colour = 'black'),
        # axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Pointrange_AUC_hits_sub_lib_glucose_umps1.pdf'),
    type = 'pdf', dpi = 300, height = 5, width = 9)










###############################
### Glucose sublibrary - TF ###
###############################



# read and transform data
glucose.tf = read_xlsx('Sub_library/Summary_TF Sublibrary_Glucose Supp_10mM_N2 worms.xlsx', sheet = 'Summary_good') 

glucose.tf = glucose.tf %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% # make table long
    unite(ID, Genes, Supplement, Replicate, remove = FALSE) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Genes = as.factor(Genes),
        ID = as.factor(ID),
        Drug = as.factor(Drug)) %>%
    select(Genes, Well, ID, Replicate, Supplement, Supplement_mM, Drug, Score)


# table with genotypes and pathways


# data summary
glu.tf.sum = glucose.tf %>%
  group_by(Supplement, Supplement_mM, Drug, Genes) %>%
  summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
  mutate(BW_Score = Median_Score[Genes == 'BW'],
         BW_Mean = Mean[Genes == 'BW']) %>%
  ungroup %>%
  group_by(Genes, Drug) %>%
  ungroup %>%
  mutate(BW_norm = Median_Score - BW_Score,
         BW_Mean_norm = Mean - BW_Mean)


### SCATTER PLOT

drug = 5
pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
glu.tf.sum %>% 
    filter(Drug == drug) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    left_join(pathways) %>% 
    ggplot(aes(x = Glucose_0, y = Glucose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(position = pos, size = 2) + 
    # geom_text_repel(aes(label = ifelse(Genes %in% c('MntR'), as.character(Genes), '')), position = pos) + # for drug 0
    geom_text_repel(aes(label = ifelse(abs(Glucose_0) >= 0.2 | abs(Glucose_10) >= 0.2 , as.character(Genes), '')), position = pos) + # for drug 1
    # geom_text_repel(aes(label = Genes), position = pos) +
    # coord_cartesian(xlim = c(-0.5, 0.5)) + # use in drug == 0
    # coord_cartesian(xlim = c(-1, 2), ylim = c(-0.4, 0.1)) + # use in drug == 1
    labs(title = expression(paste("5FU + Glucose effect on ", italic('C. elegans'), " phenotype (TFs)", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Glucose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', paste0('Scatter_tf_sub_lib_glucose_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 10, width = 12)



### Bargraphs

# all glucose barplot
glu.tf.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = glucose.tf %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    theme_light() +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'barplot_all_tf_sub_lib_glucose.pdf'),
    type = 'pdf', dpi = 300, height = 15, width = 20)



# barplot per drug concentration
drug = 5
glu.tf.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glucose.tf %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_tf_sub_lib_glucose',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 8, width = 30)





###--------------------------


# first, sum all score values by gene and condition
glu.auc = glucose.tf %>% group_by(Genes, Well, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score)) 

# filter values with NAs (maybe try to impute them with RF?)
glu.auc = glu.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


## we want to compare each gene against BW per condition
# split data
glu_0 = glu.auc %>%  filter(Supplement_mM == 0) %>% ungroup %>% select(-Well)
glu_10 = glu.auc %>% filter(Supplement_mM == 10) %>% ungroup %>% select(-Well)


# remove BWs with my dummy function
glu_0 = ctrls(glu_0)
glu_10 = ctrls(glu_10)

# extract the gene list for the loop
genes = glucose.tf %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# calculate the Mood's test (median test) for every condition
med.res = rbind(m.test(glu_0, genes), m.test(glu_10, genes))
# the same but for the t.test
ttest.res = rbind(av.test(glu_0, genes), av.test(glu_10, genes))


## after having done the statistical tests, we can filter those ones that have sig. differences
# for this dataset I will use the t test results, as the other is completely useless (data too variable, and only 2 reps)

glu.genes = ttest.res %>% filter(fdr <= 0.05) %>% select(Gene) %>% unique
glu.genes = sort(as.character(glu.genes$Gene)) ; glu.genes = c('BW', glu.genes)
glu.genes = factor(glu.genes, levels = glu.genes)

## plot hits for glucose

glu.tf.sum %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = glucose.tf %>% 
        filter(Genes %in% glu.genes) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))


quartz.save(file = here('Summary', 'Bargraph_hits_tf_lib_glucose.pdf'),
    type = 'pdf', dpi = 300, height = 13, width = 17)



# save statistics list
list_of_datasets = list('Summary statistics' = glu.tf.sum, 'AUC' = glu.auc, 't-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'glucose_TF_stats.xlsx'), colNames = T, rowNames = F) 




length(glu.genes)

# generate a small dataframe to divide the original dataframe
div = c(rep(1, 24), rep(2, 24), rep(3, 23), rep(4, 23), rep(5, 23))
df = data.frame(Genes = glu.genes, div)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, fill = Supplement_mM,  width = 0.9)) +
    geom_bar(aes(x = Genes), stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = as.character(Genes), Supplement_mM = as.character(Supplement_mM)) %>%
        left_join(df), aes(x = Genes, y = Sum), 
                            position = position_jitterdodge(jitter.width = 0.4, jitter.height = 0.05, dodge.width = 1), 
                            alpha = 0.8, show.legend = FALSE) +
    theme_light() +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Bargraph_AUC_hits_TF_lib_glucose.pdf'),
    type = 'pdf', dpi = 300, height = 10, width = 11)



# version with pointrange 

div = c(rep(1, 39), rep(2, 39), rep(3, 39))
df = data.frame(Genes = glu.genes, div)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Supplement_mM)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = factor(Genes, levels = glu.genes)) %>%
        left_join(df), aes(x = Genes, y = Sum, group = Supplement_mM), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Pointrange_AUC_hits_TF_lib_glucose.pdf'),
    type = 'pdf', dpi = 300, height = 9, width = 12)



# -------------------------------------------------------------- #
# -------------------------------------------------------------- #
# -------------------------------------------------------------- #

############################ 
### Other develop assays ###
############################


#####################################################
### gltA mutants with salvage and de novo mutants ###
#####################################################


gltA_dn_data = read_xlsx('other_develop_assays/4. DevelopAssay_denovo_salvage_gltAmuts_10mM Sugar Supp_30_01_2018_2.xlsx', sheet = 'Summary') 

gltA_dn_data = gltA_dn_data %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`, `10`, `40`) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Genes = as.factor(Genes),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Drug = as.numeric(Drug)) 


# create summary of data

gltA_dn.sum = gltA_dn_data %>%
    group_by(Supplement, Supplement_mM, Drug, Genes) %>%
    summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
    mutate(BW_score = Median_Score[Genes == 'BW'],
            BW_norm = Median_Score - BW_score) %>%
    ungroup


drug = 2.5
galac.df = gltA_dn.sum %>%
    filter(Supplement %in% c('Control', 'Galactose')) %>%
    filter(Drug == drug) %>%
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    mutate(Supplement = 'Galactose') %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm)

gluco.df = gltA_dn.sum %>%
    filter(Supplement %in% c('Control', 'Glucose')) %>%
    filter(Drug == drug) %>%
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    mutate(Supplement = 'Glucose') %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm)

glyc.df = gltA_dn.sum %>%
    filter(Supplement %in% c('Control', 'Glycerol')) %>%
    filter(Drug == drug) %>%
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    mutate(Supplement = 'Glycerol') %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm)


pos =  position_jitter(width = 0.1, height = 0.1, seed = 1) # to plot names in jitter positions
pos2 = position_jitter(width = 0.1, height = 0.1, seed = 2) # to plot names in jitter positions
pos3 = position_jitter(width = 0.1, height = 0.1, seed = 3) # to plot names in jitter positions

galac.df %>% 
    ggplot(aes(x = Galactose_0, y = Galactose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(position = pos, size = 2, colour = '#657AFF') +
    geom_point(data = gluco.df, aes(x = Glucose_0, y = Glucose_10), position = pos2, size = 2, colour = '#EB27BC') +
    geom_point(data = glyc.df, aes(x = Glycerol_0, y = Glycerol_10), position = pos3, size = 2, colour = '#DB6D18') +
    geom_text_repel(aes(label = as.character(Genes)), position = pos, colour = '#657AFF') + 
    geom_text_repel(data = gluco.df, aes(x = Glucose_0, y = Glucose_10, label = as.character(Genes)), position = pos2, box.padding = 0.5,
        colour = '#EB27BC') +
    geom_text_repel(data = glyc.df, aes(x = Glycerol_0, y = Glycerol_10, label = as.character(Genes)), position = pos3, box.padding = 1.5,
        colour = '#DB6D18') + 
    labs(title = expression(paste("5FU + Glucose effect on ", italic('C. elegans'), " phenotype ", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Glucose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', paste0('Scatter_all_glta_salvage_denovo_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 10, width = 12)



# plot heatmaps
gradcolours = c('black', 'yellow', 'orange', 'red')

diffcolours = c('purple', 'blue', 'steelblue', 'black', 'yellow', 'orange', 'red')
breaks = c(-4, -3, -2, -1, 0, 1, 2, 3, 4)
limits = c(-4 , 4)

# function to plot heatmap, from Pov
plotHeatmap = function(data, x, y, fill, grad = gradcolours, breaks = c(0,1,2,3,4), limits = c(0,4)){
  ggplot(data,aes_string(x = x, y = y, fill = fill, colour = fill))+
    geom_tile(size=0.9)+
    xlab('5FU, uM')+
    theme(axis.ticks = element_blank(),
          panel.border = element_blank(),
          panel.grid.minor = element_blank(),
          panel.grid.major = element_blank(),
          axis.line = element_line(colour = NA),
          axis.line.x = element_line(colour = NA),
          axis.line.y = element_line(colour = NA),
          strip.text = element_text(colour = 'black', face='bold',size = 10),
          axis.text.x= element_text(face = 'bold', colour = 'black', size = 10, angle = 90, hjust = 1),
          axis.text.y= element_text(face = 'bold', colour = 'black', size = 10))+
    scale_fill_gradientn(colours = grad, breaks = breaks, limits = limits, guide = "legend") +
    scale_colour_gradientn(colours = grad, breaks = breaks, limits = limits, guide = FALSE)
}



# heatmap

gltA_dn.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    plotHeatmap(x = 'Drug', y = 'Genes', fill = 'Median_Score')+
    ylab('Strain')+
    labs(fill = 'C. elegans phenotype')+
    facet_wrap(~ Supplement, ncol = 2)

quartz.save(file = here('Summary', 'Heatmap_glta_salvage_denovo_glucose.pdf'),
    type = 'pdf', dpi = 300, height = 10, width = 12)


gltA_dn.sum %>%
    ungroup %>%
    filter(Genes != 'BW') %>%
    mutate(Drug = as.factor(Drug)) %>%
    plotHeatmap(x = 'Drug', y = 'Genes', fill = 'BW_norm',grad = diffcolours, breaks = breaks, limits = limits)+
    ylab('Strain')+
    labs(fill = 'C. elegans phenotype')+
    facet_wrap(~ Supplement, ncol = 2)

quartz.save(file = here('Summary', 'Heatmap_glta_salvage_denovo_glucose_BWnorm.pdf'),
    type = 'pdf', dpi = 300, height = 10, width = 12)




#  barplot
gltA_dn.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Drug, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = gltA_dn_data %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Drug, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 1, jitter.height = 0.15), alpha = 0.8) +
    # facet_wrap(~Genes + Supplement, strip.position = 'top') +
    facet_grid(vars(Genes), vars(Supplement), scales = 'free_y') +
    coord_cartesian(ylim = c(1,4.2)) +
    # geom_vline(xintercept = 1.5, size = 0.8) +
    # scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 8, order = 3:8) +
    theme_light() +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'Bargraph_all_glta_salvage_denovo_0to40uM.pdf'),
    type = 'pdf', dpi = 300, height = 7, width = 9)



# barplot per drug concentration
drug = 40
gltA_dn.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Supplement,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = gltA_dn_data %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            # position = position_dodge2(width = 0.9), alpha = 0.8) +
                            position = position_jitterdodge(jitter.width = 0.9, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4.2)) +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_glta_salvage_denovo_sugars',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 6, width = 7)



###--------------------------
### statistical tests

# this dataset is a bit special, so a dirty way of doing this is by spliting it into 
# different datasets depending on their Supplement
ctr.auc = gltA_dn_data %>% filter(Supplement %in% c('Control')) %>% mutate(Supplement = 'Control') %>%
    group_by(Genes, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score))
galac.auc = gltA_dn_data %>% filter(Supplement %in% c('Galactose')) %>% mutate(Supplement = 'Galactose') %>%
    group_by(Genes, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score))
gluco.auc = gltA_dn_data %>% filter(Supplement %in%   c('Glucose')) %>% mutate(Supplement = 'Glucose') %>%
    group_by(Genes, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score))
glyce.auc = gltA_dn_data %>% filter(Supplement %in%  c('Glycerol')) %>% mutate(Supplement = 'Glycerol') %>%
    group_by(Genes, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score))

sugars.auc = rbind(ctr.auc, galac.auc, gluco.auc, glyce.auc)

# extract the gene list for the loop
genes = gltA_dn_data %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# t.test
ttest.res = rbind(
    av.test(ctr.auc, genes),
    av.test(galac.auc, genes),
    av.test(gluco.auc, genes),
    av.test(glyce.auc, genes))


sugars.auc %>%
    ungroup %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Supplement_mM)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = sugars.auc , aes(x = Genes, y = Sum, group = Supplement_mM), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Pointrange_AUC_gltA_sugars.pdf'),
    type = 'pdf', dpi = 300, height = 5, width = 9)



# save statistics list
list_of_datasets = list('Summary statistics' = gltA_dn.sum, 'AUC' = sugars.auc, 't-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'gltA_sugars_stats.xlsx'), colNames = T, rowNames = F) 







#--------------------------------



#################################
### glta pyruvate DMs library ###
#################################


gltA_pyr = read_xlsx('other_develop_assays/6. DevelopAssay_gltA_DMs_sublibrary_4reps_10mM_Glu_06-08-18_2.xlsx', sheet = 'Summary_good') 

gltA_pyr = gltA_pyr %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Genes = as.factor(Genes),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Drug = as.numeric(Drug)) 


# create summary of data

gltA_pyr.sum = gltA_pyr %>%
    group_by(Supplement, Supplement_mM, Drug, Genes) %>%
    summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
    mutate(BW_score = Median_Score[Genes == 'BW'],
            BW_norm = Median_Score - BW_score) %>%
    ungroup


# scatter plot

drug = 0
pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
gltA_pyr.sum %>% 
    filter(Drug == drug) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    ggplot(aes(x = Glucose_0, y = Glucose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(position = pos, size = 2) + 
    coord_cartesian(ylim = c(-1,1)) + # for drug == 0
    # geom_text_repel(aes(label = Genes), position = pos) +
    geom_text_repel(aes(label = ifelse(!Genes %in% c('Δglta ΔprpB::K','ΔgltA::K'), as.character(Genes), '')), position = pos, colour = 'black', size = 2.5) +
    geom_text_repel(aes(label = ifelse(Genes %in% c('Δglta ΔprpB::K','ΔgltA::K'), as.character(Genes), '')), position = pos, colour = 'red', size = 5, box.padding = 3.5) +
    labs(title = expression(paste("5FU + Pyruvate effect on ", italic('C. elegans'), " N2 phenotype", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' N2 phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' N2 phenotype ' , bold('(Pyruvate)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', paste0('Scatter_pyruvate_DMs_lib_glucose_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 10, width = 12)





### Bargraphs

# all glucose barplot
gltA_pyr.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = gltA_pyr %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    theme_light() +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'Bargraph_all_pyruvate_DMs_lib_glucose.pdf'),
    type = 'pdf', dpi = 300, height = 12, width = 15)



# barplot per drug concentration
drug = 5
gltA_pyr.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = gltA_pyr %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_pyruvate_DMs_lib_glucose_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 8, width = 12)




# barplot per drug concentration
drug = 5
gltA_pyr.sum %>%
    ungroup %>%
    # filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = gltA_pyr %>% 
                            # filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_pyruvate_DMs_lib_glucose_all_uM.pdf')),
    type = 'pdf', dpi = 300, height = 8, width = 25)


###--------------------------


# first, sum all score values by gene and condition
glu.auc = gltA_pyr %>% group_by(Genes, Well, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score)) 

# filter values with NAs (maybe try to impute them with RF?)
glu.auc = glu.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


## we want to compare each gene against BW per condition
# split data
glu_0 = glu.auc %>%  filter(Supplement_mM == 0) %>% ungroup %>% select(-Well)
glu_10 = glu.auc %>% filter(Supplement_mM == 10) %>% ungroup %>% select(-Well)


# remove BWs with my dummy function
glu_0 = ctrls(glu_0)
glu_10 = ctrls(glu_10)

# extract the gene list for the loop
genes = gltA_pyr %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# the same but for the t.test
ttest.res = rbind(av.test(glu_0, genes), av.test(glu_10, genes))


## after having done the statistical tests, we can filter those ones that have sig. differences
# for this dataset I will use the t test results, as the other is completely useless (data too variable, and only 2 reps)

glu.genes = ttest.res %>% filter(fdr <= 0.05) %>% select(Gene) %>% unique
glu.genes = sort(as.character(glu.genes$Gene)) ; glu.genes = c('BW', glu.genes)
glu.genes = factor(glu.genes, levels = glu.genes)

## plot hits for glucose

gltA_pyr.sum %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = gltA_pyr %>% 
        filter(Genes %in% glu.genes) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))


quartz.save(file = here('Summary', 'Bargraph_hits_pyruvate_DMs_lib_glucose_0to5uM.pdf'),
    type = 'pdf', dpi = 300, height = 13, width = 17)



# save statistics list
list_of_datasets = list('Summary statistics' = gltA_pyr.sum, 'AUC' = glu.auc, 't-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'pyruvate_DMs_stats.xlsx'), colNames = T, rowNames = F) 




length(glu.genes)

# generate a small dataframe to divide the original dataframe
div = c(rep(1, 20), rep(2, 20))
df = data.frame(Genes = glu.genes, div)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, fill = Supplement_mM,  width = 0.9)) +
    geom_bar(aes(x = Genes), stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = as.character(Genes), Supplement_mM = as.character(Supplement_mM)) %>%
        left_join(df), aes(x = Genes, y = Sum), 
                            position = position_jitterdodge(jitter.width = 0.4, jitter.height = 0.05, dodge.width = 1), 
                            alpha = 0.8, show.legend = FALSE) +
    theme_light() +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Bargraph_hits_pyruvate_DMs_lib_glucose_AUC.pdf'),
    type = 'pdf', dpi = 300, height = 10, width = 11)



# version with pointrange 
div = c(rep(1, 20), rep(2, 20))
df = data.frame(Genes = glu.genes, div)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Supplement_mM)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = factor(Genes, levels = glu.genes)) %>%
        left_join(df), aes(x = Genes, y = Sum, group = Supplement_mM), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Pointrange_hits_pyruvate_DMs_lib_glucose_AUC.pdf'),
    type = 'pdf', dpi = 300, height = 7, width = 12)





##########################################
### glta methylcitrate cycle gene DMs  ###
##########################################

gltA_mc = read_xlsx('other_develop_assays/7. DevelopAssay_glta_2-methylcitrate cycle_DMs_sublibrary_5reps_10mM_Glu_01-09-18.xlsx', sheet = 'Summary_good') 

gltA_mc = gltA_mc %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Genes = as.factor(Genes),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Drug = as.numeric(Drug)) 


# create summary of data

gltA_mc.sum = gltA_mc %>%
    group_by(Supplement, Supplement_mM, Drug, Genes) %>%
    summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
    mutate(BW_score = Median_Score[Genes == 'BW'],
            BW_norm = Median_Score - BW_score) %>%
    ungroup



# scatter plot

drug = 1
pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
gltA_mc.sum %>% 
    filter(Drug == drug) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    ggplot(aes(x = Glucose_0, y = Glucose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(position = pos, size = 2) + 
    geom_text_repel(aes(label = Genes), position = pos) +
    #coord_cartesian(ylim = c(-1,1), xlim = c(-1,1)) + # for drug == 0
    # geom_text_repel(aes(label = ifelse(Genes != 'ΔgltA::K', as.character(Genes), '')), position = pos, colour = 'black', size = 2.5) +
    # geom_text_repel(aes(label = ifelse(Genes == 'ΔgltA::K', as.character(Genes), '')), position = pos, colour = 'red', size = 5, box.padding = 3.5) +
    labs(title = expression(paste("5FU + Glucose effect on ", italic('C. elegans'), " N2 phenotype", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' N2 phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' N2 phenotype ' , bold('(Glucose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', paste0('Scatter_glta_methyl_DMs_glucose_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 10, width = 12)





### Bargraphs

# all glucose barplot
gltA_mc.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = gltA_mc %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    theme_light() +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'Bargraph_hits_glta_methyl_DMs_glucose_0to5uM.pdf'),
    type = 'pdf', dpi = 300, height = 12, width = 15)



# barplot per drug concentration
drug = 5
gltA_mc.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = gltA_mc %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_glta_methyl_DMs_glucose_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 7, width = 9)



# heatmap

gltA_mc.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    plotHeatmap(x = 'Drug', y = 'Genes', fill = 'Median_Score')+
    ylab('Strain')+
    labs(fill = 'C. elegans phenotype')+
    facet_wrap(~ Supplement_mM, ncol = 2)

quartz.save(file = here('Summary', paste0('Heatmap_gltA_methyl.pdf')),
    type = 'pdf', dpi = 300, height = 7, width = 9)


gltA_mc.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    plotHeatmap(x = 'Drug', y = 'Genes', fill = 'BW_norm',grad = diffcolours, breaks = breaks, limits = limits)+
    ylab('Strain')+
    labs(fill = 'C. elegans phenotype')+
    facet_wrap(~ Supplement_mM, ncol = 2)


quartz.save(file = here('Summary', paste0('Heatmap_gltA_methyl_BWnorm.pdf')),
    type = 'pdf', dpi = 300, height = 7, width = 9)




###--------------------------


# first, sum all score values by gene and condition
glu.auc = gltA_mc %>% group_by(Genes, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score)) 

# filter values with NAs (maybe try to impute them with RF?)
glu.auc = glu.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


## we want to compare each gene against BW per condition
# split data
glu_0 = glu.auc %>%  filter(Supplement_mM == 0) %>% ungroup 
glu_10 = glu.auc %>% filter(Supplement_mM == 10) %>% ungroup


# remove BWs with my dummy function
glu_0 = ctrls(glu_0)
glu_10 = ctrls(glu_10)

# extract the gene list for the loop
genes = gltA_mc %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# the same but for the t.test
ttest.res = rbind(av.test(glu_0, genes), av.test(glu_10, genes))


# funny dataset
ttest.res
ttest.res[is.na(ttest.res)] = 1


## after having done the statistical tests, we can filter those ones that have sig. differences
# for this dataset I will use the t test results, as the other is completely useless (data too variable, and only 2 reps)

glu.genes = ttest.res %>% filter(fdr <= 0.05) %>% select(Gene) %>% unique
glu.genes = sort(as.character(glu.genes$Gene)) ; glu.genes = c('BW', glu.genes)
glu.genes = factor(glu.genes, levels = glu.genes)

## plot hits for glucose

gltA_mc.sum %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = gltA_pyr %>% 
        filter(Genes %in% glu.genes) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))


quartz.save(file = here('Summary', 'Bargraph_hits_glta_methyl_DMs_glucose_0to5uM.pdf'),
    type = 'pdf', dpi = 300, height = 13, width = 17)



# save statistics list
list_of_datasets = list('Summary statistics' = gltA_mc.sum, 'AUC' = glu.auc, 't-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'glta_methylcitrate_stats.xlsx'), colNames = T, rowNames = F) 



length(glu.genes)

# generate a small dataframe to divide the original dataframe
div = c(rep(1, 11))
df = data.frame(Genes = glu.genes, div)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, fill = Supplement_mM,  width = 0.9)) +
    geom_bar(aes(x = Genes), stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = as.character(Genes), Supplement_mM = as.character(Supplement_mM)) %>%
        left_join(df), aes(x = Genes, y = Sum), 
                            position = position_jitterdodge(jitter.width = 0.4, jitter.height = 0.05, dodge.width = 1), 
                            alpha = 0.8, show.legend = FALSE) +
    theme_light() +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Bargraph_hits_glta_methyl_DMs_glucose_AUC.pdf'),
    type = 'pdf', dpi = 300, height = 10, width = 11)



glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Supplement_mM)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = factor(Genes, levels = glu.genes)) %>%
        left_join(df), aes(x = Genes, y = Sum, group = Supplement_mM), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Pointrange_hits_glta_methyl_DMs_glucose_AUC.pdf'),
    type = 'pdf', dpi = 300, height = 5, width = 12)







#################################################################
### overexpressing prpB, acanB and prpR in glta, prpB and DM  ###
#################################################################


# load data

mc_data = read_xlsx('other_develop_assays/9. DevelopAssay_glta_2-methylcitrate cycle_ASKA_sublibrary_5reps_10mM_Glu_13-11-18.xlsx', sheet = 'Summary_good') 


mc_data = mc_data %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Genes = as.factor(Genes),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Drug = as.numeric(Drug)) 

# create summary of data

mc.sum = mc_data %>%
    group_by(Supplement, Supplement_mM, Drug, Plasmid, Genes) %>%
    summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
    mutate(BW_score = Median_Score[Genes == 'BW'],
            BW_norm = Median_Score - BW_score) %>%
    ungroup

# generate new dataframe with BW values

mc.ctr = mc.sum %>%
    ungroup %>%
    select(Supplement:Median_Score) %>%
    filter(Genes == 'BW') %>%
    rename(BW_score = Median_Score) %>%
    select(-Genes)

mc.sum = mc.sum %>%
    ungroup %>%
    left_join(mc.ctr) %>%
    mutate(BW_norm = Median_Score - BW_score,
           Supplement = 'Glucose')

# check that the normalisation has worked
filter(mc.sum, Plasmid == 'prpB', Supplement == 'Glucose') %>% data.frame


# scatter plot
drug = 0
pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
mc.sum %>%
    filter(Drug == drug) %>% 
    select(Supplement, Supplement_mM, Genes, Plasmid, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
        ggplot(aes(x = Glucose_0, y = Glucose_10, color = Genes, shape = Plasmid)) + 
        geom_hline(yintercept = 0, colour = 'grey30') +
        geom_vline(xintercept = 0, colour = 'grey30') +
        geom_point(position = pos, size = 2.5) +    
        coord_cartesian(xlim = c(-1,1), ylim = c(-1,1)) +
        labs(title = expression(paste("ASKA library and gltA mutants effect on ", italic('C. elegans'), " phenotype, 5 uM 5FU", sep = '')),
             x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
             y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Glucose)'), sep = ' '))) +
        theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50")) + 
        guides(colour = guide_legend(override.aes = list(size = 4))) + # make lengend points larger
        geom_text_repel(aes(label = paste(Genes, Plasmid, sep = ' '), colour = Genes), position = pos)

quartz.save(file = here('Summary', paste0('Scatter_all_OE_glta_prpB_glucose_',drug,'uM.pdf')),
    type = 'pdf', dpi = 300, height = 9, width = 11)

# heatmap

mc.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    plotHeatmap(x = 'Drug', y = 'Genes', fill = 'Median_Score') +
    ylab('Strain') +
    labs(fill = 'C. elegans phenotype') +
    facet_wrap(~Supplement_mM + Plasmid, ncol = 4)

quartz.save(file = here('Summary', paste0('Heatmap_all_OE_glta_prpB_glucose_0to5uM.pdf')),
    type = 'pdf', dpi = 300, height = 6, width = 11)





### Bargraphs

# all glucose barplot
mc.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = mc_data %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    # facet_wrap(~Genes + Plasmid, strip.position = 'top') +
    facet_grid(vars(Plasmid), vars(Genes)) +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    theme_light() +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'Bargraph_all_OE_glta_prpB_glucose_0to5uM.pdf'),
    type = 'pdf', dpi = 300, height = 8, width = 10)


# barplot per drug concentration
drug = 5
mc.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = mc_data %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM + Plasmid, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_OE_glta_prpB_glucose_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 7, width = 9)





###--------------------------


# first, sum all score values by gene and condition
OE.auc = mc_data %>% group_by(Genes, Replicate, Supplement, Plasmid, Supplement_mM) %>%
    summarise(Sum = sum(Score)) 

OE.auc.sum = OE.auc %>% group_by(Genes, Supplement_mM, Plasmid, Supplement) %>%
    summarise(Mean = mean(Sum, na.rm = TRUE),
              SD = sd(Sum, na.rm = TRUE))

# filter values with NAs (maybe try to impute them with RF?)
OE.auc = OE.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


## we want to compare each gene against BW per condition
# split data
"acnB" "none" "prpB" "prpR"

OE_acnB_0 = OE.auc %>%  filter(Supplement_mM == 0 , Plasmid == 'acnB') %>% ungroup 
OE_acnB_10 = OE.auc %>% filter(Supplement_mM == 10, Plasmid == 'acnB') %>% ungroup
OE_none_0 = OE.auc %>%  filter(Supplement_mM == 0 , Plasmid == 'none') %>% ungroup 
OE_none_10 = OE.auc %>% filter(Supplement_mM == 10, Plasmid == 'none') %>% ungroup
OE_prpB_0 = OE.auc %>%  filter(Supplement_mM == 0 , Plasmid == 'prpB') %>% ungroup 
OE_prpB_10 = OE.auc %>% filter(Supplement_mM == 10, Plasmid == 'prpB') %>% ungroup
OE_prpR_0 = OE.auc %>%  filter(Supplement_mM == 0 , Plasmid == 'prpR') %>% ungroup 
OE_prpR_10 = OE.auc %>% filter(Supplement_mM == 10, Plasmid == 'prpR') %>% ungroup


# extract the gene list for the loop
genes = mc_data %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# the same but for the t.test
ttest.res1 = rbind(av.test(OE_acnB_0, genes), av.test(OE_acnB_10, genes)) %>% mutate(Plasmid = 'acnB')
ttest.res2 = rbind(av.test(OE_none_0, genes), av.test(OE_none_10, genes)) %>% mutate(Plasmid = 'none')
ttest.res3 = rbind(av.test(OE_prpB_0, genes), av.test(OE_prpB_10, genes)) %>% mutate(Plasmid = 'prpB')
ttest.res4 = rbind(av.test(OE_prpR_0, genes), av.test(OE_prpR_10, genes)) %>% mutate(Plasmid = 'prpR')

ttest.res = rbind(ttest.res1,ttest.res2,ttest.res3,ttest.res4)

# funny dataset
ttest.res
ttest.res[is.na(ttest.res)] = 1


# save statistics list
list_of_datasets = list('Summary statistics' = mc.sum, 'AUC' = OE.auc, 'AUC_sum' = OE.auc.sum,'t-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'OE_glta_prpB_stats.xlsx'), colNames = T, rowNames = F) 






############################################################
### pyruvate supplementation on glta and methyl mutants  ###
############################################################

pyr_meth = read_xlsx('other_develop_assays/10. DevelopAssay_glta_2-methylcitrate cycle_sublibrary_5reps_10mM_Pyruvate_13-11-18_2.xlsx', sheet = 'Summary_good') 

pyr_meth = pyr_meth %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Genes = as.factor(Genes),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Drug = as.numeric(Drug)) 


# create summary of data

pyr_meth.sum = pyr_meth %>%
    group_by(Supplement, Supplement_mM, Drug, Genes) %>%
    summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
    mutate(BW_score = Median_Score[Genes == 'BW'],
            BW_norm = Median_Score - BW_score) %>%
    ungroup



# scatter plot

drug = 5
pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
pyr_meth.sum %>% 
    filter(Drug == drug) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    ggplot(aes(x = Pyruvate_0, y = Pyruvate_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(position = pos, size = 2) + 
    geom_text_repel(aes(label = Genes), position = pos) +
    # coord_cartesian(ylim = c(-1,1), xlim = c(-1,1)) + # for drug == 0
    # geom_text_repel(aes(label = ifelse(Genes != 'ΔgltA::K', as.character(Genes), '')), position = pos, colour = 'black', size = 2.5) +
    # geom_text_repel(aes(label = ifelse(Genes == 'ΔgltA::K', as.character(Genes), '')), position = pos, colour = 'red', size = 5, box.padding = 3.5) +
    labs(title = expression(paste("5FU + Pyruvate effect on ", italic('C. elegans'), " N2 phenotype", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' N2 phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' N2 phenotype ' , bold('(Pyruvate)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', paste0('Scatter_methyl_muts_pyruvate_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 10, width = 12)





### Bargraphs

# all glucose barplot
pyr_meth.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = pyr_meth %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    theme_light() +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'Bargraph_all_methyl_muts_pyruvate_0to5uM.pdf'),
    type = 'pdf', dpi = 300, height = 12, width = 15)



# barplot per drug concentration
drug = 5
pyr_meth.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = pyr_meth %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_methyl_muts_pyruvate_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 7, width = 9)



# heatmap

pyr_meth.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    plotHeatmap(x = 'Drug', y = 'Genes', fill = 'Median_Score')+
    ylab('Strain')+
    labs(fill = 'C. elegans phenotype')+
    facet_wrap(~ Supplement_mM, ncol = 2)

quartz.save(file = here('Summary', paste0('Heatmap_methyl_muts_pyruvate.pdf')),
    type = 'pdf', dpi = 300, height = 7, width = 9)


pyr_meth.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    plotHeatmap(x = 'Drug', y = 'Genes', fill = 'BW_norm',grad = diffcolours, breaks = breaks, limits = limits)+
    ylab('Strain')+
    labs(fill = 'C. elegans phenotype')+
    facet_wrap(~ Supplement_mM, ncol = 2)


quartz.save(file = here('Summary', paste0('Heatmap_methyl_muts_pyruvate_BWnorm.pdf')),
    type = 'pdf', dpi = 300, height = 7, width = 9)




###--------------------------


# first, sum all score values by gene and condition
glu.auc = pyr_meth %>% group_by(Genes, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score)) 

glu.auc.sum = glu.auc %>% group_by(Genes, Supplement_mM, Supplement) %>%
    summarise(Mean = mean(Sum, na.rm = TRUE),
              SD = sd(Sum, na.rm = TRUE))


# filter values with NAs (maybe try to impute them with RF?)
glu.auc = glu.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


## we want to compare each gene against BW per condition
# split data
glu_0 = glu.auc %>%  filter(Supplement_mM == 0) %>% ungroup 
glu_10 = glu.auc %>% filter(Supplement_mM == 10) %>% ungroup


# remove BWs with my dummy function
glu_0 = ctrls(glu_0)
glu_10 = ctrls(glu_10)

# extract the gene list for the loop
genes = pyr_meth %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# the same but for the t.test
ttest.res = rbind(av.test(glu_0, genes), av.test(glu_10, genes))


# funny dataset
ttest.res
ttest.res[is.na(ttest.res)] = 1


## after having done the statistical tests, we can filter those ones that have sig. differences
# for this dataset I will use the t test results, as the other is completely useless (data too variable, and only 2 reps)

glu.genes = ttest.res %>% filter(fdr <= 0.05) %>% select(Gene) %>% unique
glu.genes = sort(as.character(glu.genes$Gene)) ; glu.genes = c('BW', glu.genes)
glu.genes = factor(glu.genes, levels = glu.genes)

## plot hits for glucose

pyr_meth.sum %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = pyr_meth %>% 
        filter(Genes %in% glu.genes) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))


quartz.save(file = here('Summary', 'Bargraph_hits_methyl_muts_pyruvate_0to5uM.pdf'),
    type = 'pdf', dpi = 300, height = 13, width = 17)



# save statistics list
list_of_datasets = list('Summary statistics' = gltA_mc.sum, 'AUC' = glu.auc, 'AUC_sum' = glu.auc.sum,'t-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'methyl_muts_pyruvate_stats.xlsx'), colNames = T, rowNames = F) 



length(glu.genes)

# generate a small dataframe to divide the original dataframe
div = c(rep(1, 13))
df = data.frame(Genes = glu.genes, div)

glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, fill = Supplement_mM,  width = 0.9)) +
    geom_bar(aes(x = Genes), stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = as.character(Genes), Supplement_mM = as.character(Supplement_mM)) %>%
        left_join(df), aes(x = Genes, y = Sum), 
                            position = position_jitterdodge(jitter.width = 0.4, jitter.height = 0.05, dodge.width = 1), 
                            alpha = 0.8, show.legend = FALSE) +
    theme_light() +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Bargraph_hits_methyl_muts_pyruvate_AUC.pdf'),
    type = 'pdf', dpi = 300, height = 6, width = 11)



glu.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Supplement_mM)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = factor(Genes, levels = glu.genes)) %>%
        left_join(df), aes(x = Genes, y = Sum, group = Supplement_mM), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Pointrange_hits_methyl_muts_pyruvate_AUC.pdf'),
    type = 'pdf', dpi = 300, height = 5, width = 12)









##############################################
### overexpressing prpR on all methyl muts ###
##############################################


# load data

prpR = read_xlsx('other_develop_assays/19. DevelopAssay_glta_2-methylcitrate cycle_ASKA_prpR_OE_sublibrary_3reps_03-05-19.xlsx', sheet = 'Summary_good') 


prpR = prpR %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Genes = as.factor(Genes),
        Drug = as.numeric(Drug)) 

# create summary of data

prp.sum = prpR %>%
    group_by(Drug, Plasmid, Genes) %>%
    summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
    mutate(BW_score = Median_Score[Genes == 'BW'],
            BW_norm = Median_Score - BW_score) %>%
    ungroup

# generate new dataframe with BW values

mc.ctr = prp.sum %>%
    ungroup %>%
    select(Drug:Median_Score) %>%
    filter(Genes == 'BW') %>%
    rename(BW_score = Median_Score) %>%
    select(-Genes)

prp.sum = prp.sum %>%
    ungroup %>%
    left_join(mc.ctr) %>%
    mutate(BW_norm = Median_Score - BW_score)

# heatmap

prp.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    plotHeatmap(x = 'Drug', y = 'Genes', fill = 'Median_Score') +
    ylab('Strain') +
    labs(fill = 'C. elegans phenotype') +
    facet_wrap(~Plasmid, ncol = 2)

quartz.save(file = here('Summary', paste0('Heatmap_OE_prpR_methyl_muts.pdf')),
    type = 'pdf', dpi = 300, height = 6, width = 11)



### Bargraphs

# all glucose barplot
prp.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Plasmid, y = Median_Score, fill = interaction(Plasmid, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = prpR %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Plasmid, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    # facet_grid(vars(Plasmid), vars(Genes)) +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    theme_light() +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'Bargraph_all_OE_prpR_methyl_muts_0to5uM .pdf'),
    type = 'pdf', dpi = 300, height = 8, width = 10)


# barplot per drug concentration
drug = 5
prp.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = prpR %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Plasmid, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_OE_prpR_methyl_muts_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 7, width = 9)





###--------------------------


# first, sum all score values by gene and condition
OE.auc = prpR %>% group_by(Genes, Replicate, Plasmid) %>%
    summarise(Sum = sum(Score)) 

OE.auc.sum = OE.auc %>% group_by(Genes, Plasmid) %>%
    summarise(Mean = mean(Sum, na.rm = TRUE),
              SD = sd(Sum, na.rm = TRUE))

# filter values with NAs (maybe try to impute them with RF?)
OE.auc = OE.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


## we want to compare each gene against BW per condition
# split data


OE_prpR = OE.auc %>%  filter(Plasmid == 'prpR') %>% ungroup 
OE_none = OE.auc %>%  filter(Plasmid == 'None') %>% ungroup


# extract the gene list for the loop
genes = prpR %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)


# function that calculates the comparison between BW and other genes in the datasets
# lm version!
av.test2 = function(data, genes) {
    results = data.frame()
    for (i in 1:length(genes)){
        sub = data %>% filter(Genes %in% c('BW', genes[i]))
        if(dim(sub)[1] <= 2) {next} # skips genes that are not present (kinda)
        model = lm(Sum ~ Genes, data = sub)
        df = tidy(model)[2,]
        df['Gene'] = genes[i]
        df['Plasmid'] = unique(data$Plasmid)
        results = rbind(results,df)
    }
    results = results %>% mutate(p.stars = stars.pval(p.value), 
            fdr = p.adjust(p.value, method = 'fdr'), fdr.stars = stars.pval(fdr)) %>%
        select(Gene, Plasmid, estimate, std.error, statistic, p.value, p.stars, fdr, fdr.stars)
    return(results)
}


# the same but for the t.test
ttest.res = rbind(av.test2(OE_prpR, genes), av.test2(OE_none, genes)) 
# funny dataset
ttest.res
ttest.res[is.na(ttest.res)] = 1


# save statistics list
list_of_datasets = list('Summary statistics' = prp.sum, 'AUC' = OE.auc, 'AUC_sum' = OE.auc.sum,'t-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'OE_prpR_methyl_muts__stats.xlsx'), colNames = T, rowNames = F) 



###

glu.genes = ttest.res %>% filter(fdr <= 0.05) %>% select(Gene) %>% unique
glu.genes = sort(as.character(glu.genes$Gene)) ; glu.genes = c('BW', glu.genes)
glu.genes = factor(glu.genes, levels = glu.genes)

## plot hits 

prp.sum %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Plasmid, y = Median_Score, fill = interaction(Plasmid, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = prpR %>% 
        filter(Genes %in% glu.genes) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Plasmid, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    # scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'))


quartz.save(file = here('Summary', 'Bargraph_hits_OE_prpR_methyl_mutsto5uM.pdf'),
    type = 'pdf', dpi = 300, height = 13, width = 17)


length(glu.genes)

# generate a small dataframe to divide the original dataframe
div = c(rep(1, 7))
df = data.frame(Genes = glu.genes, div)

OE.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Plasmid) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, fill = Plasmid,  width = 0.9)) +
    geom_bar(aes(x = Genes), stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = OE.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = as.character(Genes), Plasmid = as.character(Plasmid)) %>%
        left_join(df), aes(x = Genes, y = Sum), 
                            position = position_jitterdodge(jitter.width = 0.4, jitter.height = 0.05, dodge.width = 1), 
                            alpha = 0.8, show.legend = FALSE) +
    theme_light() +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Bargraph_hits_OE_prpR_methyl_muts_AUC.pdf'),
    type = 'pdf', dpi = 300, height = 6, width = 11)



OE.auc %>%
    ungroup %>%
    filter(Genes %in% glu.genes) %>%
    mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Plasmid) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Plasmid)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = OE.auc %>% 
        filter(Genes %in% glu.genes) %>% ungroup %>%
        mutate(Genes = factor(Genes, levels = glu.genes)) %>%
        left_join(df), aes(x = Genes, y = Sum, group = Plasmid), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Pointrange_hits_OE_prpR_methyl_muts_AUC.pdf'),
    type = 'pdf', dpi = 300, height = 5, width = 12)






#################################################################
### supplementing gltA and methyl muts with Methylisocitrate  ###
#################################################################

methiso = read_xlsx('other_develop_assays/20. DevelopAssay_methylcitrate cycle muts_3reps_10mM_2-Methyisocitric_acid_03-06-19.xlsx', sheet = 'Summary_good') 

methiso = methiso %>%
    gather(Drug, Score, `0`, `1`, `2.5`, `5`) %>% 
    mutate(Replicate = as.numeric(Replicate),
        Genes = as.factor(Genes),
        Supplement_mM = as.factor(Supplement_mM),
        Supplement = as.factor(Supplement),
        Drug = as.numeric(Drug)) 


# create summary of data

methiso.sum = methiso %>%
    group_by(Supplement, Supplement_mM, Drug, Genes) %>%
    summarise(Median_Score = median(Score, na.rm = TRUE),
            MAD = mad(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
    mutate(BW_score = Median_Score[Genes == 'BW'],
            BW_norm = Median_Score - BW_score) %>%
    ungroup



# scatter plot

drug = 5
pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions
methiso.sum %>% 
    filter(Drug == drug) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    ggplot(aes(x = Methylisocitrate_0, y = Methylisocitrate_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(position = pos, size = 2) + 
    geom_text_repel(aes(label = Genes), position = pos) +
    # coord_cartesian(ylim = c(-1,1), xlim = c(-1,1)) + # for drug == 0
    # geom_text_repel(aes(label = ifelse(Genes != 'ΔgltA::K', as.character(Genes), '')), position = pos, colour = 'black', size = 2.5) +
    # geom_text_repel(aes(label = ifelse(Genes == 'ΔgltA::K', as.character(Genes), '')), position = pos, colour = 'red', size = 5, box.padding = 3.5) +
    labs(title = expression(paste("5FU + 2-Methyisocitrate effect on ", italic('C. elegans'), " N2 phenotype", sep = '')),
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' N2 phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' N2 phenotype ' , bold('(2-Methyisocitrate)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
            panel.grid.major = element_line(colour = "grey90"),
            panel.background = element_rect(fill = "white", colour = "grey50"),
            legend.text = element_text(size = 6)) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', paste0('Scatter_methyl_muts_methylisocit_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 10, width = 12)





### Bargraphs

# all glucose barplot
methiso.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Supplement_mM, y = Median_Score, fill = interaction(Supplement_mM, Drug),  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position=position_dodge(.9)) +
    geom_point(data = methiso %>% 
                            mutate(Drug = as.factor(Drug)), aes(x = Supplement_mM, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Genes, strip.position = 'top') +
    coord_cartesian(ylim = c(1,4)) +
    geom_vline(xintercept = 1.5, size = 0.8) +
    scale_fill_manual(values = c('#8ABDCE','#BF82A6','#3ACDFF','#8E4E74', '#009BD0','#6B214C','#003749','#51193A'), name = 'Supplement - Drug') +
    theme_light() +
    labs(x = 'Supplement (in mM)',
         y = 'Median') +
    theme(strip.text = element_text(colour = 'black'))

quartz.save(file = here('Summary', 'Bargraph_all_methyl_muts_methylisocit_0to5uM.pdf'),
    type = 'pdf', dpi = 300, height = 12, width = 15)



# barplot per drug concentration
drug = 5
methiso.sum %>%
    ungroup %>%
    filter(Drug == drug) %>%
    mutate(Drug = as.factor(Drug)) %>%
    ggplot(aes(x = Genes, y = Median_Score, fill = Drug,  width = 0.9)) +
    geom_bar(stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(ymin = Median_Score - MAD, ymax = Median_Score + MAD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = methiso %>% 
                            filter(Drug == drug) %>%
                            mutate(Drug = as.factor(Drug)), aes(x = Genes, y = Score, group = Drug), 
                            position = position_jitterdodge(jitter.width = 0.25, jitter.height = 0.05), alpha = 0.8) +
    facet_wrap(~Supplement_mM, strip.position = 'top', nrow = 2) +
    coord_cartesian(ylim = c(1,4)) +
    scale_fill_discrete_sequential(palette = "Blues", nmax = 6, order = 3:6) +
    theme_light() +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1))

quartz.save(file = here('Summary', paste0('Bargraph_all_methyl_muts_methylisocit_',as.character(drug),'uM.pdf')),
    type = 'pdf', dpi = 300, height = 6, width = 7)



# heatmap

methiso.sum %>%
    ungroup %>%
    mutate(Drug = as.factor(Drug)) %>%
    plotHeatmap(x = 'Drug', y = 'Genes', fill = 'Median_Score')+
    ylab('Strain')+
    labs(fill = 'C. elegans phenotype')+
    facet_wrap(~ Supplement_mM, ncol = 2)

quartz.save(file = here('Summary', paste0('Heatmap_methyl_muts_methylisocit.pdf')),
    type = 'pdf', dpi = 300, height = 7, width = 9)



###--------------------------


# first, sum all score values by gene and condition
glu.auc = methiso %>% group_by(Genes, Replicate, Supplement, Supplement_mM) %>%
    summarise(Sum = sum(Score)) 

glu.auc.sum = glu.auc %>% group_by(Genes, Supplement_mM, Supplement) %>%
    summarise(Mean = mean(Sum, na.rm = TRUE),
              SD = sd(Sum, na.rm = TRUE))


# filter values with NAs (maybe try to impute them with RF?)
glu.auc = glu.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


## we want to compare each gene against BW per condition
# split data
glu_0 = glu.auc %>%  filter(Supplement_mM == 0) %>% ungroup 
glu_10 = glu.auc %>% filter(Supplement_mM == 10) %>% ungroup


# remove BWs with my dummy function
glu_0 = ctrls(glu_0)
glu_10 = ctrls(glu_10)

# extract the gene list for the loop
genes = methiso %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
genes = as.character(genes$Genes)

# the same but for the t.test
ttest.res = rbind(av.test(glu_0, genes), av.test(glu_10, genes))


## after having done the statistical tests, we can filter those ones that have sig. differences
# for this dataset I will use the t test results, as the other is completely useless (data too variable, and only 2 reps)

glu.genes = ttest.res %>% filter(fdr <= 0.05) %>% select(Gene) %>% unique
glu.genes = sort(as.character(glu.genes$Gene)) ; glu.genes = c('BW', glu.genes)
glu.genes = factor(glu.genes, levels = glu.genes)


# save statistics list
list_of_datasets = list('Summary statistics' = methiso.sum, 'AUC' = glu.auc, 'AUC_sum' = glu.auc.sum,'t-student test' = ttest.res)

write.xlsx(list_of_datasets, here('Summary', 'methyl_muts_methyliso_stats.xlsx'), colNames = T, rowNames = F) 



length(genes)

# generate a small dataframe to divide the original dataframe
div = c(rep(1, 4))
df = data.frame(Genes = genes, div)

glu.auc %>%
    ungroup %>%
    # filter(Genes %in% genes) %>%
    # mutate(Genes = factor(Genes, levels = genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, fill = Supplement_mM,  width = 0.9)) +
    geom_bar(aes(x = Genes), stat = "identity", position = position_dodge2(), colour = 'black') +
    geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        # filter(Genes %in% genes) %>% ungroup %>%
        # mutate(Genes = as.character(Genes), Supplement_mM = as.character(Supplement_mM)) %>%
        left_join(df), aes(x = Genes, y = Sum), 
                            position = position_jitterdodge(jitter.width = 0.4, jitter.height = 0.05, dodge.width = 1), 
                            alpha = 0.8, show.legend = FALSE) +
    theme_light() +
        labs(x = 'Genes',
             y = 'AUC mean') +
    # facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Bargraph_all_methyl_muts_methylisocit_AUC.pdf'),
    type = 'pdf', dpi = 300, height = 6, width = 7)



glu.auc %>%
    ungroup %>%
    # filter(Genes %in% glu.genes) %>%
    # mutate(Genes = factor(Genes, levels = glu.genes)) %>%
    group_by(Genes, Supplement_mM) %>%
    summarise(Mean = mean(Sum), 
                SD = sd(Sum)) %>%
    left_join(df) %>%
    ungroup %>%
    ggplot(aes(x = Genes, y = Mean, width = 0.9, colour = Supplement_mM)) +
    # geom_errorbar(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), width = 0.2, position = position_dodge(.9)) +
    geom_point(data = glu.auc %>% 
        # filter(Genes %in% glu.genes) %>% ungroup %>%
        # mutate(Genes = factor(Genes, levels = glu.genes)) %>%
        left_join(df), aes(x = Genes, y = Sum, group = Supplement_mM), 
                            position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0.05, dodge.width = 0.4), 
                            show.legend = FALSE) +
    geom_pointrange(aes(x = Genes, ymin = Mean - SD, ymax = Mean + SD), stat = "identity", position = position_dodge2(width = 0.4), 
         alpha = 1, shape = '-', size = 0.8, fatten = 11) +
    theme_light() +
    scale_color_manual(values = c('black', '#FC1C32')) +
        labs(x = 'Genes',
             y = 'AUC mean') +
    # facet_wrap(~div, scales = "free_x", nrow = 5) +
    theme(strip.text = element_text(colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.background = element_blank(),  # delete facet_wrap 
        strip.text.x = element_blank())         # delete facet_wrap 

quartz.save(file = here('Summary', 'Pointrange_all_methyl_muts_methylisocit_AUC.pdf'),
    type = 'pdf', dpi = 300, height = 5, width = 12)












################################################################
### BIG screen purine and oxidative phosphorylation analysis ###
################################################################




### load the data

# use this dataset
glucose 

# load the file with the pathways. Remember there are 3 version, the last one (kegg3)
pathways = read_xlsx(here('Sub_library/big_screen','EcoCyc_pathways_KeioSublibrary_08-12-18.xlsx'), 
    sheet = 'paths')

pathways = pathways %>% 
    mutate(Genes = as.factor(Genes),
           Pathway = as.factor(Pathway),
           KEGG1 = as.factor(KEGG1),
           KEGG2 = as.factor(KEGG2),
           KEGG3 = as.factor(KEGG3))

# summary of the data

sublib.sum = glucose %>%
  group_by(Supplement, Supplement_mM, Drug, Genes) %>%
  summarise(Median_Score = median(Score, na.rm = TRUE),
            Mean = mean(Score, na.rm = TRUE),
            SD = sd(Score, na.rm = TRUE)) %>%
  mutate(BW_Score = Median_Score[Genes == 'BW'],
         BW_Mean = Mean[Genes == 'BW']) %>%
  ungroup %>%
  group_by(Genes, Drug) %>%
  mutate(No_supplement = ifelse("0" %in% Supplement_mM, Median_Score[Supplement_mM == '0'], NA)) %>% # Get phenotype values at Supplement_mM=0 
  ungroup %>%
  mutate(BW_norm = Median_Score - BW_Score,
         BW_Mean_norm = Mean - BW_Mean,
         Suppl_norm = Median_Score - No_supplement,
         Interaction = Median_Score - BW_Score - No_supplement,
         Supplement = 'Glucose')


# range of colours
colours17 = c('#000000','#8C158C','#8C158C','#0A2CFF','#006006',
            '#FF008C','#9D4300','#7B008C','#008BF6','#5EAA44',
            '#FF1500','#A42900','#006A54','#00CDFF','#F0BF8F',
            '#B5B0B8','#FAE603')
colours_kegg2 = c('#F0BF8F','#000000','#8C158C','#0A2CFF','#006006',
                  '#008BF6','#5EAA44','#FF1500','#CC80E6','#BAE303')
colours_kegg3 = c('#8A7090','#000000','#53BAE2','#EF6713','#58A32D','#EF1613','#F45A95','#8C5A3A')


pos = position_jitter(width = 0.05, height = 0.05, seed = 1) # to plot names in jitter positions

# test if the data is OK

sublib.sum %>% 
    left_join(pathways) %>%
    filter(Drug == 5) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm, Pathway:KEGG3) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    ggplot(aes(x = Glucose_0, y = Glucose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(aes(colour = KEGG3), position = pos, size = 2) + 
    scale_color_manual(values = colours_kegg3) + 
    scale_fill_manual(values = colours_kegg3) +
    # geom_text_repel(aes(label = ifelse(Pathway == 'ribonucleotides de novo biosynthesis', as.character(Genes), '')), position = pos)
    geom_text_repel(aes(label = Genes, colour = KEGG3), position = pos) +
    labs(title = "5FU + Supplement (Glucose) effect on different BW mutants",
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Glucose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
        panel.grid.major = element_line(colour = "grey90"),
        panel.background = element_rect(fill = "white", colour = "grey50")) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger


###
### load the KEIO library data


# load data
## REMEMBER, PLATE A AND PLATE B ARE NOW 97 AND 99
keiolib = read_xlsx(here('Sub_library/big_screen','Keio Library.xlsx'), sheet = 'ALL_final') 

keiolib = keiolib %>%
    filter(!is.na(Gene), !(Gene %in% c('WT', 'present', 'no bact' )))


## running some checks

# is there any gene repeated? 

subset = keiolib %>% filter(Gene != 'Control')
length(unique(subset$Gene)) == length(subset$Gene)
# if TRUE, everything is fine



# transform data into 1 column data frame
keiolib = keiolib %>% 
    gather(Sample, Score, Control_1:GluFU_3) %>%
    separate(Sample, c('Sample', 'Replicate'), sep = '_') %>%
    mutate_at(c('Plate', 'Column', 'Row', 'Gene', 'Sample'), as.factor) %>%
    mutate(Score = as.numeric(Score))


# Controls median
controls = keiolib %>%
    filter(Gene == 'Control') %>%
    mutate(Score = as.numeric(Score)) %>%
    group_by(Sample) %>%
    summarise(Median = median(Score))

# we can use the values of the first row of the Controls as the median, it has the same values
controls = filter(keiolib, Gene == 'Control', Sample == 'Control')[1,]
controls = rbind(controls, filter(keiolib, Gene == 'Control', Sample == 'FU')[1,])
controls = rbind(controls, filter(keiolib, Gene == 'Control', Sample == 'Glu')[1,])
controls = rbind(controls, filter(keiolib, Gene == 'Control', Sample == 'GluFU')[1,])


## summarise the data
keio.sum = keiolib %>%
    filter(Gene != 'Control') %>%
    rbind(., controls) %>%
    group_by(Sample, Gene) %>%
    summarise(Median_Score = median(Score, na.rm = TRUE),
              Mean = mean(Score, na.rm = TRUE),
              SD = sd(Score, na.rm = TRUE)) %>%
    mutate(BW_Score = Median_Score[Gene == 'Control'],
           BW_Mean = Mean[Gene == 'Control']) %>%
    ungroup %>%
    mutate(BW_norm = Median_Score - BW_Score,
           BW_Mean_norm = Mean - BW_Mean,
           Supplement = 'Glucose')

# adding a helper column
keio.sum['Supplement_mM'] = 0
keio.sum[keio.sum$Sample == 'Glu',]$Supplement_mM = 10
keio.sum[keio.sum$Sample == 'GluFU',]$Supplement_mM = 10



### test data if everything is OK
pos = position_jitter(width = 0.15, height = 0.15, seed = 1) # to plot names in jitter positions
keio.sum %>% 
    filter(Sample %in% c('FU', 'GluFU')) %>%
    select(Supplement, Supplement_mM, Gene, BW_norm) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    ggplot(aes(x = Glucose_0, y = Glucose_10)) +
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(position = pos, size = 2, alpha = .2) +
    # geom_text_repel(aes(label = ifelse(Glucose_0 > 1.5 & abs(Glucose_10) > 0.5, as.character(Gene), '')), position = pos) +
    # geom_text_repel(aes(label = ifelse((abs(Glucose_0) > 1.5 | abs(Glucose_10) > 0.5), as.character(Gene), '')), position = pos) +
    labs(title = "5FU + Supplement (Glucose) effect on different BW mutants",
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Glucose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
        panel.grid.major = element_line(colour = "grey90"),
        panel.background = element_rect(fill = "white", colour = "grey50")) 



####
## now load the info from the supplementary material from Leo's previous work


# load data
supp = read_xlsx(here('Sub_library/big_screen','mmc2.xlsx'), sheet = 'Enrichment_with_genes') 

# oxidative phosphorylation 
ox_res = supp %>% 
    select(Category, Term_ID, Term, Gene, Plate) %>%
    filter(Term == 'Oxidative phosphorylation')

# purine metabolism
pur_met = supp %>% 
    select(Category, Term_ID, Term, Gene, Plate) %>%
    filter(Term == 'Purine metabolism')




# is any of these genes in the original sublibrary?
intersect(unique(sublib$Genes), ox_res$Gene)
# there are four genes found to be repeated in the sublibrary: "sdhA", "sdhB", "sdhC", "sdhD"

# subset the keiolib with the genes from the oxidative phosphorilation

keio.genes = keio.sum %>% 
    mutate(Gene = as.character(Gene)) %>%
    filter(Gene %in% ox_res$Gene) %>% 
    mutate(Gene = recode(Gene, sdhA = 'sdhA_K',
                               sdhB = 'sdhB_K',
                               sdhC = 'sdhC_K',
                               sdhD = 'sdhD_K'), 
           Drug = 0,
           Drug = ifelse(Sample %in% c('FU', 'GluFU'), 5, 0),
           Pathway = 'Oxidative phosphorylation') %>%
    rename(Genes = Gene) %>%
    select(-Sample)

keio.pur.genes = keio.sum %>% 
    mutate(Gene = as.character(Gene)) %>%
    filter(Gene %in% pur_met$Gene) %>% 
    mutate(Gene = recode(Gene, pykA = 'pykA_K',
                               pgm = 'pgm_K',
                               nrdF = 'nrdF_K',
                               nrdE = 'nrdE_K',
                               nrdD = 'nrdD_K',
                               mazG = 'mazG_K',
                               ndk = 'ndk_K',
                               yjjG = 'yjjG_K',
                               pykF = 'pykF_K'),
           Drug = 0,
           Drug = ifelse(Sample %in% c('FU', 'GluFU'), 5, 0),
           Pathway = 'Purine metabolism') %>%
    rename(Genes = Gene) %>%
    select(-Sample)




# join both datasets
join = sublib.sum %>% 
    select(Genes, Median_Score, Mean, SD, BW_Score, BW_Mean, BW_norm, BW_Mean_norm, Supplement, Supplement_mM, Drug) %>%
    left_join(pathways[,c(1,5)]) %>% 
    rename(Pathway = KEGG3) %>%
    rbind(keio.genes, keio.pur.genes)



joined_colours = c('#8A7090','#000000','#53BAE2','#EF6713','#58A32D','#EF1613','#F45A95','#8C5A3A', 'grey50', '#FF00DC')

join %>% 
    filter(Drug == 5) %>% 
    select(Supplement, Supplement_mM, Genes, BW_norm, Pathway:Pathway) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    ggplot(aes(x = Glucose_0, y = Glucose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(aes(colour = Pathway), position = pos, size = 2) + 
    scale_color_manual(values = joined_colours) + 
    scale_fill_manual(values = joined_colours) +
    # geom_text_repel(aes(label = ifelse(Pathway == 'ribonucleotides de novo biosynthesis', as.character(Genes), '')), position = pos)
    geom_text_repel(aes(label = Genes, colour = Pathway), position = pos) +
    labs(title = "5FU + Supplement (Glucose) effect on different BW mutants",
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Glucose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
        panel.grid.major = element_line(colour = "grey90"),
        panel.background = element_rect(fill = "white", colour = "grey50")) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger

quartz.save(file = here('Summary', 'sublibrary_extra_pathways.pdf'),
    type = 'pdf', dpi = 300, height = 10, width = 12)




# only the two new pathways with gltA and pyrE
join %>% 
    filter(Drug == 5, Pathway %in% c('Oxidative phosphorylation', 'Purine metabolism')) %>% 
    rbind(., join %>% filter(Drug == 5, Genes %in% c('pyrE', 'gltA'))) %>%
    select(Supplement, Supplement_mM, Genes, BW_norm, Pathway:Pathway) %>%
    unite(Supp, Supplement, Supplement_mM) %>%
    spread(Supp, BW_norm) %>%
    ggplot(aes(x = Glucose_0, y = Glucose_10)) + 
    geom_hline(yintercept = 0, colour = 'grey30') +
    geom_vline(xintercept = 0, colour = 'grey30') +
    geom_point(aes(colour = Pathway), position = pos, size = 2) + 
    scale_color_manual(values = c('#BF112E', '#C7046D', '#A86200', '#0A2DF5')) + 
    scale_fill_manual(values = c('#BF112E', '#C7046D', '#A86200', '#0A2DF5')) +
    # geom_text_repel(aes(label = ifelse(Pathway == 'ribonucleotides de novo biosynthesis', as.character(Genes), '')), position = pos)
    geom_text_repel(aes(label = Genes, colour = Pathway), position = pos) +
    labs(title = "5FU + Supplement (Glucose) effect on different BW mutants",
         x = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype', sep = ' ')),
         y = expression(paste('Normalised median scores of ', italic('C. elegans'), ' phenotype ' , bold('(Glucose)'), sep = ' '))) +
    theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"),
        panel.grid.major = element_line(colour = "grey90"),
        panel.background = element_rect(fill = "white", colour = "grey50")) + 
    guides(colour = guide_legend(override.aes = list(size = 4))) # make lengend points larger



quartz.save(file = here('Summary', 'sublibrary_extra_pathways_purine.pdf'),
    type = 'pdf', dpi = 300, height = 10, width = 12)




## statistics 
# preparation of data

# keio.ctr = keiolib %>%
#     mutate(Gene = as.character(Gene), 
#            Gene = recode(Gene, 'Control' = 'BW'),
#            Drug = 0,
#            Drug = ifelse(Sample %in% c('FU', 'GluFU'), 5, 0),
#            Pathway = 'Control') %>%
#     filter(Gene == 'BW') %>%
#     rename(Genes = Gene) %>%
#     select(-Sample)

# controls
keio.ctr = controls %>% 
    mutate(Gene = as.character(Gene), 
           Gene = recode(Gene, 'Control' = 'BW'),
           Drug = 0,
           Supplement_mM = 10,
           Drug = ifelse(Sample %in% c('FU', 'GluFU'), 5, 0),
           Supplement_mM = ifelse(Sample %in% c('Glu', 'GluFU'), 10, 0),
           Pathway = 'Control') %>%
    rename(Genes = Gene) %>%
    select(-Sample)

keio.ctr2 = keio.ctr %>% mutate(Replicate = 2)
keio.ctr3 = keio.ctr %>% mutate(Replicate = 3)


# purines
keio.pur = keiolib %>% 
    mutate(Gene = as.character(Gene)) %>%
    filter(Gene %in% pur_met$Gene) %>% 
    mutate(Drug = 0,
           Supplement_mM = 10,
           Drug = ifelse(Sample %in% c('FU', 'GluFU'), 5, 0),
           Supplement_mM = ifelse(Sample %in% c('Glu', 'GluFU'), 10, 0),
           Pathway = 'Purine metabolism') %>%
    rename(Genes = Gene) %>%
    select(-Sample)

# ox phosphorylation
keio.ox = keiolib %>% 
    mutate(Gene = as.character(Gene)) %>%
    filter(Gene %in% ox_res$Gene) %>% 
    mutate(
           Drug = 0,
           Supplement_mM = 10,
           Drug = ifelse(Sample %in% c('FU', 'GluFU'), 5, 0),
           Supplement_mM = ifelse(Sample %in% c('Glu', 'GluFU'), 10, 0),
           Pathway = 'Oxidative phosphorylation') %>%
    rename(Genes = Gene) %>%
    select(-Sample)

keio.sub = rbind(keio.ctr, keio.ctr2, keio.ctr3, keio.pur, keio.ox)



keio.sum.pur = keio.sum %>% 
    filter(Gene %in% keio.sub$Genes) %>%
    mutate(Gene = as.character(Gene)) %>%  
    left_join(keio.sub %>% rename(Gene = Genes) %>% 
        select(Gene, Pathway)) %>%
    unique(.)



# first, sum all score values by gene and condition
keio.auc = keio.sub %>% group_by(Genes, Replicate, Supplement_mM) %>%
    summarise(Sum = sum(Score)) 



# save statistics list
list_of_datasets = list('Summary statistics' = keio.sum.pur, 'AUC' = keio.auc)

write.xlsx(list_of_datasets, here('Summary', 'KEIO_purine_oxphosphorylation.xlsx'), colNames = T, rowNames = F) 





# # filter values with NAs (maybe try to impute them with RF?)
# keio.auc = keio.auc %>% filter_at(vars(Sum), any_vars(!is.na(.))) 


# ## we want to compare each gene against BW per condition
# # split data
# glu_0 = keio.auc %>%  filter(Supplement_mM == 0) 
# glu_10 = keio.auc %>% filter(Supplement_mM == 10)







# # # remove BWs with my dummy function
# # glu_0 = ctrls(glu_0)
# # glu_10 = ctrls(glu_10)

# # extract the gene list for the loop
# genes = glucose.tf %>% ungroup %>% filter(Genes != 'BW') %>% select(Genes) %>% unique %>% data.frame 
# genes = as.character(genes$Genes)

# # calculate the Mood's test (median test) for every condition
# med.res = rbind(m.test(glu_0, genes), m.test(glu_10, genes))
# # the same but for the t.test
# ttest.res = rbind(av.test(glu_0, genes), av.test(glu_10, genes))


# ## after having done the statistical tests, we can filter those ones that have sig. differences
# # for this dataset I will use the t test results, as the other is completely useless (data too variable, and only 2 reps)

# glu.genes = ttest.res %>% filter(fdr <= 0.05) %>% select(Gene) %>% unique
# glu.genes = sort(as.character(glu.genes$Gene)) ; glu.genes = c('BW', glu.genes)
# glu.genes = factor(glu.genes, levels = glu.genes)
























########################################################################
########################################################################
####################### Plots for QQ Poster ############################
########################################################################
########################################################################


###
### 4-way plot
###

# read contrast adjustments, it seems there is a bug in some part of the
# code and it's repeating the same values from only condition into the others

adjustments = as_tibble(read.csv('Contrast_adjustments.csv'))

# is a hell of a code, it's super messy, but it works...

worm.sum[worm.sum$Strain == 'CRP', ]$Strain = 'crp'
worm.sum$Strain = as.factor(worm.sum$Strain)



# print the list of variables 
for (name in unique(worm.sum$Strain)){
  print(name)
  print(unique((worm.sum %>% filter(Strain == name))$Experiment))
}

# select only one experiment and strain
# contrasts and other main variables of this part

strain = 'BW'
experiment = "T5"


str_C = paste(strain, '_C', sep = '')
str_T = paste(strain, '_T', sep = '')
str_CT = paste(strain, '_5FU', sep = '')
alln = paste(strain, '_5FU_Alln', sep = '')
adjs = paste(strain, '_5FU_adj', sep = '')


# for example, to mimic Pov's code
worm.sum %>% filter(Experiment == experiment, Strain == strain)

worm.old = worm.sum %>%
  filter(Experiment == experiment & Strain == strain) %>% 
  ungroup %>%
  mutate(Description = 'C. elegans development at 5uM 5-FU',
         Contrast = 'Ce_Dev5',   # C. elegans developement with 5FU
         Contrast_type = 'Treatment',
         FDR = NA) %>%
  select(Description, Contrast, Contrast_type, Plate, Well, logFC = Median, SE = SD, FDR)

# check that we filtered stuff properly
worm.old %>%
  group_by(Plate) %>%
  summarise(N = n())


# Join bacterial and worm results (change contrasts)
jointresults = allresults$results %>%
  filter((Contrast %in% c(str_C, str_T, str_CT, alln))) %>%
  mutate(Contrast = fct_recode(Contrast, adjs = alln)) %>% 
  select(Description, Contrast, Contrast_type, Plate, Well, logFC, SE, FDR) %>%
  bind_rows(worm.old) %>%
  left_join(info) %>%
  select(Description:Well,Index:KEGG_ID,logFC:FDR)


jointcast = jointresults %>%
  select(Contrast, Plate, Well, Index, Metabolite, MetaboliteU, EcoCycID, KEGG_ID, logFC, SE, FDR) %>%
  gather(Stat, Value, logFC, SE, FDR) %>%
  unite(CS, Contrast, Stat) %>%
  spread(CS, Value) %>%
  rename(Ce_Dev5_Median = Ce_Dev5_logFC, Ce_Dev5_SD = Ce_Dev5_SE) %>%
  select(-Ce_Dev5_FDR)


# write.csv(jointresults,paste(odir,'/Ecoli_results_All_As_old_screen.csv', sep = ''),row.names = FALSE)
# write.csv(jointcast,paste(odir,'/Ecoli_results_sidebyside_As_old_screen.csv', sep = ''),row.names = FALSE)

 
# Multiplex 
jointresults.multi = multiplex(jointresults,c("Plate","Well","Index","Metabolite","MetaboliteU","EcoCycID","KEGG_ID"))
jointresults.multi2 = multiplex(jointresults,c("Plate","Well","Index","Metabolite","MetaboliteU","EcoCycID","KEGG_ID"),2)




########################
NGMa = adjustments[adjustments$Contrast == alln, ]$a
NGMb = adjustments[adjustments$Contrast == alln, ]$b



gradcolours = c('#71B83B','yellow','orange','red')




total = jointresults.multi %>%
  filter(x_Contrast == str_C & y_Contrast == str_T & z_Contrast == 'Ce_Dev5')

first = jointresults.multi %>%
  filter(x_Contrast == str_C & y_Contrast == str_T & z_Contrast == 'Ce_Dev5') %>%
  filter(z_logFC <=1)

second = jointresults.multi %>%
  filter(x_Contrast == str_C & y_Contrast == str_T & z_Contrast == 'Ce_Dev5') %>%
  filter(z_logFC >1 & z_logFC <= 2)

third = jointresults.multi %>%
  filter(x_Contrast == str_C & y_Contrast == str_T & z_Contrast == 'Ce_Dev5') %>%
  filter(z_logFC >2 & z_logFC <= 3)

fourth = jointresults.multi %>%
  filter(x_Contrast == str_C & y_Contrast == str_T & z_Contrast == 'Ce_Dev5') %>%
  filter(z_logFC >3 & z_logFC <= 4)


ggplot(first, aes()) +
  geom_abline(intercept = 0, slope = 1, alpha = 1, color = 'grey', linetype = 'longdash') +
  geom_abline(aes(intercept = NGMb, slope = NGMa), alpha = 0.8, color = 'red') +
  geom_vline(aes(xintercept = 0), alpha = 0.9, color = 'grey') +
  geom_hline(aes(yintercept = 0), alpha = 0.9, color = 'grey') +
  geom_errorbarh(data = total, aes(y = y_logFC, xmax = x_logFC + x_SE, xmin = x_logFC - x_SE), height = 0, alpha = 0.3, color = 'grey50') +
  geom_errorbar(data = total, aes(x = x_logFC, ymax = y_logFC + y_SE, ymin = y_logFC - y_SE), width = 0, alpha = 0.3, color = 'grey50') +
  geom_point(aes(x = x_logFC, y = y_logFC, colour = z_logFC), alpha = 0.85, size = 4) +
  geom_point(data = second, aes(x = x_logFC, y = y_logFC, colour = z_logFC), alpha = 0.85, size = 4) +
  geom_point(data = third, aes(x = x_logFC, y = y_logFC, colour = z_logFC), alpha = 0.85, size = 4) +
  geom_point(data = fourth, aes(x = x_logFC, y = y_logFC, colour = z_logFC), alpha = 0.85, size = 4) +
  scale_size(range = c(1,5), name = 'C. elegans\nphenotype') +
  scale_colour_gradientn(colours = gradcolours,
                         breaks = c(1,2,3,4), limits = c(1,4), guide = "legend", name = 'C. elegans\nphenotype') +
  scale_y_continuous(breaks = -5:5)+
  coord_cartesian(xlim = c(-5, 2), ylim = c(-4, 4)) +
  labs(title = paste(strain, " growth with 5FU treatment at 5 uM", sep = ''),
  x = expression(paste(italic("E. coli"), ' growth vs NGM - Control, logFC', sep = '')), 
  y = expression(paste(italic('E. coli'), ' growth vs NGM - 5-FU Treatment, logFC', sep = ''))) +
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
  # geom_text_repel(data = fourth, aes(x = x_logFC, y = y_logFC, label = ifelse(MetaboliteU %in% mlevels, MetaboliteU, '')), size = 4, box.padding = unit(1.6, "lines")) + # USE ONLY FOR DEMONSTRATIVE PURPOSES FOR LEO
  geom_text_repel(data = fourth, aes(x = x_logFC, y = y_logFC, label = ifelse(z_logFC >= 4, MetaboliteU, '')), box.padding = unit(0.2, "lines"), segment.alpha = 0.6, size = 5.5) + # only name those nutr with C. elegans phenotype 3 or more
  # geom_text_repel(data = third, aes(x = x_logFC, y = y_logFC, label = ifelse(z_logFC >= 3, MetaboliteU, '')), box.padding = unit(0.6, "lines"), segment.alpha = 0.4) + # only name those nutr with C. elegans phenotype 3 or more
  # geom_text_repel(aes(label = ifelse(z_logFC >= 3 | z_logFC == 0, MetaboliteU, '')), box.padding = unit(0.6, "lines"), segment.alpha = 0.4) + # only name those nutr with C. elegans phenotype 3 or more
  guides(color = guide_legend()) +
  theme(panel.grid.major = element_line(size = 0.06, colour = "grey50"),
      # panel.grid.major.y = element_blank(),
      panel.grid.minor.y = element_blank(),
      panel.background = element_rect(fill = "white", colour = "grey50"),
      legend.text = element_text(size = 15),
      legend.title = element_text(size = 15),
      axis.text.x = element_text(size = 10),
      axis.text.y = element_text(size = 10)) +
  guides(colour = guide_legend(override.aes = list(size = 4)))


quartz.save(file = here('Summary', paste0("QQPOSTER_Scatter_", strain, "_5uM.pdf")),
  type = 'pdf', dpi = 300, height = 8, width = 10)


# quartz.save(file = here('Summary', paste0("Scatter_EXAMPLE_", strain, "_5uM.pdf")),
#     type = 'pdf', dpi = 300, height = 8, width = 10)


geom_text_repel(aes(label = ifelse(Genes %in% c('Δglta ΔprpB::K','ΔgltA::K'), as.character(Genes), '')), position = pos, colour = 'red', size = 5, box.padding = 3.5) +

# scales for different mutants:
# BW: coord_cartesian(xlim = c(-5, 2), ylim = c(-4, 4))
# pyrE: coord_cartesian(xlim = c(-5, 2), ylim = c(-4, 4))
# TM: coord_cartesian(xlim = c(-4, 2), ylim = c(-4, 2.5))


































